---
title: "Hierarchical-based bacterial population genomics analysis using R"
author: "Gomes-Neto JC, Pavlovikj N, Benson AK"
output:
  word_document: default
  html_document: default
---

This case study will be done using a Salmonella Newport dataset that is available on NCBI-SRA, and contains
2,365 genomes. 

A list of genomes and datasets are all available here:
https://figshare.com/account/home#/projects/116625
This link to Figshare requires login credentials. 

How to install all packages. Packages should be installed only once. From time-to-time new versions will be 
available and the user is responsible for updating them accordingly. Make sure versions of programs are reported every time you use them. 

When beginning running this script, you can remove the # that comes before the install.packages() function, but when running it the second time around, and subsequently, comment the function out using #. That way you avoid creating issues with dependencies, and different versions of the program. 

```{r, include = TRUE}
# install tidyverse 
# install.packages("tidyverse")
# install skimr 
# install.packages("skimr")
# install vegan 
# install.packages("vegan")
# install forcats
# install.packages("forcats")
# install naniar
# install.packages("naniar")
# install ggpubr
# install.packages("ggpubr")
# install ggrepel
# install.packages("ggrepel")
# install reshape2
# install.packages("reshape2")
# install reshape2
# install.packages("RColorBrewer")
# install ggtree 
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ggtree")
# installation of ggtree will prompt a question about installation - answer is "a" to install/update all dependencies 
```

How to activate packages prior to utilization.

```{r, include = TRUE}
# Load previously installed packages
library(tidyverse)
library(skimr)
library(vegan)
library(forcats)
library(naniar)
library(ggtree)
library(ggpubr)
library(ggrepel)
library(reshape2)
library(RColorBrewer)
```

Enter and quality control all genotypic data including: serovar-predictions (generated by SISTR), BAPS level 1 (generated by fastbaps), ST lineages (generated by mlst), 
and cgMLST variants (generated by SISTR). All input files were generated by the describe programs which are part of the computational platform called ProkEvo.

```{r, include = TRUE}
# enter the BAPS data 
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# check the first six rows of the dataset
head(baps)
# changing the first two column names because for hierarchical analysis we only use BAPS1
colnames(baps)[1:2] <- c("id", "baps1") 
# check the first six rows of the datset again to see the change in column names
head(baps)
# select columns id and baps_1
baps1 <- baps %>% 
              select(id, baps1)
# check the first six rows of the dataset
head(baps1)
# quality control baps1 data 
skim(baps1)
# using a plotting strategy to check for missing values
vis_miss(baps1)
# rename BAPS level 1 sub-groups or haplotypes 
baps1$baps_1 <- ifelse(baps1$baps1 == 1, "BAPS1 sub-group 1", # the ifelse functions tests for different conditions prior to assigning groups to one category or another 
                   ifelse(baps1$baps1 == 2, "BAPS1 sub-group 2", 
                          ifelse(baps1$baps1 == 3, "BAPS1 sub-group 3", 
                                 ifelse(baps1$baps1 == 4, "BAPS1 sub-group 4", 
                                        ifelse(baps1$baps1 == 5, "BAPS1 sub-group 5", 
                                               ifelse(baps1$baps1 == 6, "BAPS1 sub-group 6", 
                                                       ifelse(baps1$baps1 == 7, "BAPS1 sub-group 7", 
                                                               ifelse(baps1$baps1 == 8, "BAPS1 sub-group 8", "BAPS1 sub-group 9"))))))))
##########################################--------############################################
#########################################--------#############################################
# enter MLST results
mlst <- read_csv('salmonellast_output.csv')
# check the first six rows of the dataset
head(mlst)
# generate the id column by deriving it from the FILE column
mlst$id <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# check the first six rows of the dataset
head(mlst)
# select the id and ST columns 
mlst1 <- mlst %>%
          select(id, ST) # select id and ST columns 
# use the table() function to detect extraneous characters such as "-" or "?" in the data - those are ST misclassifications
# check for the presence of ST misclassification in the ST column 
table(mlst1$ST)
# check for missing values
skim(mlst1)
# mutate the "-" character to NA (missing value)
mlst1 <- mlst1 %>%
            mutate_all(na_if, "-") # fill all - with NAs
# check the first six rows of the dataset
head(mlst1)
# quality control baps1 data 
skim(mlst1)
# using a plotting strategy to check for missing values
vis_miss(mlst1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# Enter SISTR results
sistr <- read_csv('sistr_output.csv')
# check the first six rows of the dataset
head(sistr)
# generate the id column by deriving it from the genome column
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select the id and cgmlst_ST columns 
sistr1 <- sistr %>%
            select(id, serovar_cgmlst, cgmlst_ST) # select id, serovar_cgmlst, and cgmlst_ST columns 
# check the first six rows of the dataset
head(sistr1)
# quality control baps1 data 
skim(sistr1)
# using a plotting strategy to check for missing values
vis_miss(sistr1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# combine baps1, mlst1, and sistr1 datasets
d1 <- left_join(baps1, mlst1, on = "id") # merge datasets on identical ids 
d2 <- left_join(d1, sistr1, on = "id") # merge datasets on identical ids 
# check data dimensionality 
dim(d2)
# check the first six rows of the dataset
head(d2)
# quality control baps1 data 
skim(d2)
# using a plotting strategy to check for missing values
vis_miss(d2)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# group all genomes not classified as Newport as "Other serovars"
# create a new column called serovar that contains the binary classification to group all SISTR-based misclassified genomes
d2$serovar <- ifelse(d2$serovar_cgmlst == "Newport", "Newport",
                            "Other serovars") # classify serovar_cgmlst into Newport or Others 
# check the first six rows of the dataset
head(d2)
# check the composition of the serovar column 
table(d2$serovar)
##########################################--------############################################
#########################################--------#############################################
# no transformation is needed for the baps1 dataset - no groupings or aggregations are needed 
##########################################--------############################################
#########################################--------#############################################
# check ST distribution to focus on major STs for all subsequent analyses by grouping by ST and counting
st_dist <- d2 %>%
  group_by(ST) %>% # group by the ST column
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange the counts in decreasing order 
st_dist
# based on the frequency analysis, the following STs were not aggregrated: ST118, ST45, ST5, ST132, ST31, ST350, and ST46
# create a new st column
d2$st <- ifelse(d2$ST == 5, "ST5", # create a new ST column for which minor STs are aggregated as Others 
                   ifelse(d2$ST == 31, "ST31", 
                          ifelse(d2$ST == 45, "ST45", 
                                 ifelse(d2$ST == 46, "ST46", 
                                        ifelse(d2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(d2$ST == 350, "ST350", "Other STs")))))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# check cgMLST distribution by ST to focus on major cgMLSTs for all subsequent analyses by grouping by STs and cgMLSTs and counting
cgmlst_dist <- d2 %>%
  group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange in descending order 
cgmlst_dist
# for the purposes of this analysis only the top four most frequent cgMLSTs were selected and respectively renamed
d2 <- mutate(d2, cgmlst = ifelse(cgmlst_ST %in% 1468400426, "cgMLST 1468400426", # create a new cgMLST column while aggregating minor cgMLST variants 
                                     ifelse(cgmlst_ST %in% 88443731, "cgMLST 88443731", 
                                            ifelse(cgmlst_ST %in% 1271156802, "cgMLST 1271156802",
                                                 ifelse(cgmlst_ST %in% 3336043520, "cgMLST 3336043520",  
                                            "Other cgMLSTs")))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# select only needed columns for all subsequent analyses 
d3 <- d2 %>% 
          select(id, serovar, baps_1, st, cgmlst) # select columns of interest which are described within parenthesis 
# check data dimensionality to make sure it matches that of d2 
dim(d3)
# check the first six rows of the dataset
head(d3)
# quality control baps1 data 
skim(d3)
# using a plotting strategy to check for missing values
vis_miss(d3)
# make sure there is no NA in the dataset
sum(is.na(d3))
# replace NA in the ST column
d3 <- d3 %>% mutate(st = replace_na(st, "Other STs")) 
# make sure there is no NA in the dataset
sum(is.na(d3))
# check the distribution of serovars, baps1, st, and cgmlst classifications
table(d3$serovar)
table(d3$baps_1)
table(d3$st)
table(d3$cgmlst)
# check the first six rows in the dataset 
head(d3)
# check the last six rows in the dataset 
tail(d3)
# check for missing values
vis_miss(d3)
# check the data characteristics 
str(d3)
```

Figure 1. Visualize the phylogeny-based hierarchical distribution of genotypes.

At the center of the figure is the core-genome phylogeny. Genotypic information is plotted onto it in the following order:
Serovar (innermost)
BAPS1
ST
cgMLST (outermost)

```{r, include = TRUE}
# enter the phylogeny data
tree <- read.tree("newport_phylogeny.tree")
# recall the metadata combining all hierarchical genotypes 
# here I assign d3 to d4 just to keep the formatted metadata read for other steps, in case I need to do 
# any transformation with the data to plot with the phylogenetic tree
d4 <- d3
# to plot with the phylogeny using ggtree we need to make the id column into index first 
d4 <- column_to_rownames(d4, var = "id") 
# create the tree
# adjusting the parameters to make the tree visible or however you wish requires some trial and error 
tree_plot <- ggtree(tree, layout = "circular") + xlim(-250, NA)
# plot figure 1 - color scheme for each layer of the plot should be chosen based on the user preferences 
figure_1 <- gheatmap(tree_plot, d4, offset=.0, width=20, colnames = FALSE) +  # visualize the tree with metadata 
  scale_fill_manual(values = c("coral", "darkblue",
                               "cornflowerblue", "coral", "purple", "red", "brown", "darkseagreen3", "darkblue", "darkgreen", "yellow",
                               "orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue", "black", "coral",
                               "black", "darkgreen", "purple", "darkblue", "gray"),
                    breaks = c("Newport", "Other serovars",
                               "BAPS1 sub-group 1", "BAPS1 sub-group 2", "BAPS1 sub-group 3", "BAPS1 sub-group 4", 
                               "BAPS1 sub-group 5", "BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 8", 
                               "BAPS1 sub-group 9",
                               "ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs",
                               "cgMLST 1468400426", "cgMLST 1271156802", "cgMLST 3336043520", "cgMLST 88443731", "Other cgMLSTs"),
                    name="Serovar (innermost) -> BAPS1 -> ST -> cgMLST (outermost)") +  # add colors and labels for the scale
    ggtitle(expression(bold(paste("Hierarchical population structure - ", bolditalic(S),". Newport")))) +  # add title for the figure
  theme(plot.title = element_text(size = 40, face = "bold"), 
        legend.title=element_text(size=24, face = "bold"), 
    legend.text=element_text(size=22))  # customize figure's title, legend, font
figure_1
```

Figure 2. Visualize the frequency-based distribution of hierarchical genotypes. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# bring back the metadata file
d5 <- d3 
##########################################--------############################################
#########################################--------#############################################
# plot Serovar distribution
# calculate the relative frequencies for serovar
# drop the serovars with NAs, group by serovar and then calculate the frequencies
serovar_data <- d5 %>%
              drop_na(serovar) %>% # remove NA based on serovar
              select(serovar) %>% # select the serovar column
              group_by(serovar) %>% # group_by serovar
              summarise(n = n()) %>% # count the number of genomes per serovar classification
              mutate(prop_serovar = n/sum(n)*100) # calculate the proportion of each serovar
# order serovar groups 
serovar_data$serovar <- factor(serovar_data$serovar, levels=c("Other serovars", "Newport"))
# plot the data 
sero_plot <- ggplot(serovar_data, aes(x = prop_serovar, y = serovar)) +  # show serovars on y-axis and proportion on x-axis
  ylab("") + xlab("Proportion") + xlim(0, 100) +  # set labels for axis and limit for x-axis
  ggtitle("A") +  # add title for the plot
  theme_bw() + # choose the plot background
  theme(legend.position = "none") + # remove legend
  theme(axis.text.y = element_text(size = 30)) + # set the font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # set the font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # set the font size for x-axis title
  theme(axis.text.x = element_text(angle = 0, hjust = 0.7, size = 26)) + # set the font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue")  # fill the bars that represent the values with blue
##########################################--------############################################
#########################################--------#############################################
# plot BAPS1 sub-group distribution for Newport data only
# calculate the relative frequencies for serovar and BAPS1 sub-group
# drop the serovars and BAPS1 sub-groups records with NAs, group by serovar and BAPS1 sub-group and then calculate the frequencies
baps_data <- d5 %>%
              drop_na(serovar, baps_1) %>% # drop NAs for serovar and baps_1 columns
              select(serovar, baps_1) %>% # select serovar and baps_1 columns
              group_by(serovar, baps_1) %>% # group data based on serovar and baps_1
              summarise(n = n()) %>% # count the number of observations per group
              mutate(prop = n/sum(n)*100) %>% # calculate the propotion for each group
              filter(serovar != "Other serovars") # filter out genomes classified as Other serovars 
# re-order the baps_1 column
baps_data$baps_1 <- factor(baps_data$baps_1, levels=c("BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 4", "BAPS1 sub-group 1", "BAPS1 sub-group 8"))
# plot the data 
baps_plot <- ggplot(baps_data, aes(x = baps_1, y = prop)) +  # show BAPS sub-groups on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 105) +  # set labels for axis and limit for y-axis
  ggtitle("B") +  # add title for the plot
  theme_bw() + # setting plot background
  theme(legend.position = "none") + # remove legend 
  theme(axis.text.y = element_text(size = 30)) + # change font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change font size for x-axis title 
  theme(axis.text.x = element_text(angle = 0, size = 30)) + # change font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font 
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# plot ST distribution
# calculate the relative frequencies for serovar and ST
# drop the serovars and STs with NAs, group by serovar and ST and then calculate the frequencies
st_data <- d5 %>%
              drop_na(serovar, st) %>% # drop NAs for serovar and st columns 
              select(serovar, st) %>% # select serovar and st columns 
              group_by(serovar, st) %>% # group data based on serovar and st columns 
              summarise(n = n()) %>% # count the number of observations 
              mutate(prop_st = n/sum(n)*100) %>% # calculate the proportion by groups 
              filter(serovar != "Other serovars") # filter out data for Other serovars 

# re-order the st column
st_data$st <- factor(st_data$st, levels=c("Other STs", "ST46", "ST350", "ST31", "ST132", "ST5", "ST45",
                                          "ST118"))
# plot the data
st_plot <- ggplot(st_data, aes(x = st, y = prop_st)) +  # show STs on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 105) +  # set labels for axis and limit for y-axis
  ggtitle("C") +  # add title for the plot
  theme_bw() + # set the plot background 
  theme(legend.position = "none") + # remove legends 
  theme(axis.text.y = element_text(size = 28)) + # change font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change font size for x-axis title 
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 28)) + # change font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# plot cgMLST distribution
# calculate the relative frequencies for serovar and cgMLST
# drop the serovars and cgMLSTs with NAs, group by serovar and cgMLST and then calculate the frequencies
cgmlst_data <- d5 %>%
              drop_na(serovar, cgmlst) %>% # drop NAs for serovar and cgmlst columns 
              group_by(serovar, cgmlst) %>% # group data based on serovar and cgmlst columns 
              summarise(n = n()) %>% # count the number of observations 
              mutate(prop_cgmlst = n/sum(n)*100) %>% # calculate the proportions 
              filter(serovar != "Other serovars") # filter out data for Other serovars 

# re-order cgmlst column
cgmlst_data$cgmlst <- factor(cgmlst_data$cgmlst, levels=c("Other cgMLSTs", "cgMLST 88443731", "cgMLST 3336043520", "cgMLST 1271156802", "cgMLST 1468400426"))
# plot the data 
cgmlst_plot <- ggplot(cgmlst_data, aes(x = cgmlst, y = prop_cgmlst)) +  # show cgMLSTs on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  ggtitle("D") +  # add title for the plot
  theme_bw() + # set plot background 
  theme(legend.position = "none") + # remove legend 
  theme(axis.text.y = element_text(size = 28)) + # set the font size for y-axis text 
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # set the font size for x-axis title
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 28)) + # set the font size for x-axis text, orientations, and angle 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# combine all plots in one to make Figure 2 based on 2 columns x 2 rows 
figure_2  <- ggarrange(sero_plot, baps_plot, st_plot, cgmlst_plot,
                  nrow = 2, ncol = 2, widths = c(10, 10, 10, 10),
                heights = c(1, 1, 1, 1))
figure_2
```

Supplementary Figure S1. Scatter plot showing the distribution of STs and cgMLSTs. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# bring the data containing the ST and cgmlst_ST information as numeric values
# assigning d2 to d2b
d2b <- d2
# plot the ST distribution using a scatter plot
# select Newport serovars only, drop the STs with NAs, group by ST and then calculate the distribution
st_scatter <- d2b %>%
              filter(serovar == "Newport") %>% # filter for Newport data only
              select(ST) %>% # select the ST column
              mutate(ST = as.numeric(ST)) %>% # transform the column to numeric 
              drop_na(ST) %>% # remove NA
              group_by(ST) %>% # group by ST 
              summarise(n = n()) %>% # count observations 
              mutate(prop = n/sum(n)*100) %>% # create a column with proportions 
              arrange(desc(prop))  # arrange data in descending order 
# check the first six observation of st_scatter
head(st_scatter)
# plot 
st_plot <- ggplot(st_scatter, aes(x = ST, y = prop, labels = ST)) +  # show STs on x-axis and proportion on y-axis
  xlab("ST lineages") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  ggtitle("A") +  # add title for the plot
  theme_bw() + # set plot background 
  theme(legend.position = "none") + # remove legends 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 28)) + # change x-axis text font size, angle and orientation
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize figure's title, legend, font
  geom_point(aes(size =  prop), color = "steelblue") +  # the points that represent the values are blue with size based on the proportion
  geom_text_repel(data=subset(st_scatter, prop > 1), aes(label = ST, size = 30), hjust = -6)  # add text/proportion to the plot
st_plot
##########################################--------############################################
#########################################--------#############################################
# plot the cgMLST distribution using a scatter plot
# select Newport serovars only, drop the cgMLSTs with NAs, group by cgMLST and then calculate the distribution
cgmlst_scatter <- d2b %>%
              filter(serovar == "Newport") %>% # select Newport genomes only
              select(cgmlst_ST) %>% # select the cgmlst_ST column
              mutate(cgmlst_ST = as.numeric(cgmlst_ST)) %>% # change column to numeric 
              drop_na(cgmlst_ST) %>% # drop NAs
              group_by(cgmlst_ST) %>% # group by cgmlst_ST
              summarise(n = n()) %>% # count observations 
              mutate(prop = n/sum(n)*100) %>% # calculate proportions 
              arrange(desc(prop)) # arrange in descending order 
# check the first six observation of st_scatter
head(cgmlst_scatter)
# plot 
cgmlst_plot <- ggplot(cgmlst_scatter, aes(x = cgmlst_ST, y = prop, labels = cgmlst_ST)) +  # show cgMLSTs on x-axis and proportion on y-axis
  xlab("cgMLST variants") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  ggtitle("B") +  # add title for the plot
  theme_bw() + # set plot background
  theme(legend.position = "none") + # remove legends 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 28)) + # change x-axis text font size, angle, and orientation
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize figure's title, legend, font
  geom_point(aes(size =  prop), color = "steelblue") +  # the points that represent the values are blue with size based on the proportion
  geom_text_repel(data=subset(cgmlst_scatter, prop > 3), aes(label = cgmlst_ST, size = 30), hjust = 2)  # add text/proportion to the plot
cgmlst_plot
##########################################--------############################################
#########################################--------#############################################
# combine all plots in one to make Figure 3
sup_fig_s1  <- ggarrange(st_plot, cgmlst_plot,
                  nrow = 1, ncol = 2, widths = c(10, 10),
                heights = c(1, 1)) # combine plots together in one row and two columns
sup_fig_s1
```

Figure 3. Distribution of STs based on BAPS level 1 sub-groups. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# bring the data containing the ST information as numeric values
# assigning d2 to d2b
d2b <- d2
# plot the ST distribution using a scatter plot
# select Newport serovars only, drop the STs and BAPS1 sub-groups with NAs, group by ST and BAPS1 sub-groups and then calculate the distribution
baps <- d2b %>%
              filter(serovar == "Newport") %>% # filter Newport serovars 
              select(baps_1, ST) %>% # select baps_1 and ST columns 
              mutate(ST = as.numeric(ST)) %>% # change ST column to numeric 
              drop_na(baps_1, ST) %>% # drop NAs 
              group_by(baps_1, ST) %>% # group by baps_1 and ST 
              summarise(n = n()) %>% # count observations 
              mutate(prop = n/sum(n)*100) # calculate proportions 
# check the first six observation of baps
head(baps)
# plot 
figure_3 <- ggplot(baps, aes(x = ST, y = prop, labels = ST)) +  # show STs on x-axis and proportion on y-axis
  xlab("ST lineages") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  theme_bw() + # set plot background 
  theme(legend.position = "none") + # remove legends 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 28)) + # change x-axis text font size, angle, and orientation
  theme(strip.text.x = element_text(size = 30, colour = "black", angle = 0)) +  # customize figure's title, legend, font
  geom_point(aes(size =  prop), color = "steelblue") +  # the points that represent the values are blue with size based on the proportion
  geom_text_repel(data=subset(baps, prop > 1), aes(label = ST, size = 70), hjust = -10) +  # add text/proportion to the plot
  facet_wrap(~ baps_1)  # generate multi-plots for BAPS1 sub-groups
figure_3
```

Supplementary Figure S2. Distribution of cgMLSTs based on the major ST lineages presented in Figures 1 and 2. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# bring the data containing the cgMLST information as numeric values
# assigning d2 to d2b
d2b <- d2
# plot the ST distribution using a scatter plot
# select Newport serovars only, drop the STs and cgMLSTs with NAs, group by ST and cgMLST and then calculate the distribution
cgmlst <- d2b %>%
              filter(serovar == "Newport") %>% # filter for Newport genomes only
              select(st, cgmlst_ST) %>% # select st and cgmlst_ST columns 
              drop_na(st, cgmlst_ST) %>% # drop NAs 
              group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST 
              summarise(n = n()) %>% # count observations 
              mutate(prop = n/sum(n)*100) # calculate proportions 
# check the first six observation of cgmlst
head(cgmlst)
# re-order ST 
cgmlst$st <- factor(cgmlst$st, levels=c("ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs"))
# plot 
sup_fig_2 <- ggplot(cgmlst, aes(x = cgmlst_ST, y = prop, labels = cgmlst_ST)) +  # show cgMLSTs on x-axis and proportion on y-axis
  xlab("cgMLST variants") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  theme_bw() + # set plot background 
  theme(legend.position = "none") + # remove legend 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 80, hjust = 1, size = 28)) + # change x-axis text font size, angle, and orientation
  theme(strip.text.x = element_text(size = 30, colour = "black", angle = 0)) +  # customize figure's title, legend, font
  geom_point(aes(size =  prop), color = "steelblue") +  # the points that represent the values are blue with size based on the proportion
  geom_text_repel(data=subset(cgmlst, prop > 15), aes(label = cgmlst_ST, size = 50), hjust = -25) +  # add text/proportion to the plot
  facet_wrap(~ st)  # generate multi-plots for STs
sup_fig_2
```

Supplementary Figure S3. Simpson's D diversity analysis of ST lineages using cgMLST and BAPS levels 1-6. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# enter BAPS1-6 data
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# changing column names 
colnames(baps)[1:7] <- c("id", "BAPS1", "BAPS2", "BAPS3", "BAPS4", "BAPS5", "BAPS6") 
# assign baps data to d1
data1 <- baps
# transform BAPS sub-groups to factor (categorical data)
data1 <- data1 %>% mutate(BAPS1 = as.factor(BAPS1))
data1 <- data1 %>% mutate(BAPS2 = as.factor(BAPS2))
data1 <- data1 %>% mutate(BAPS3 = as.factor(BAPS3))
data1 <- data1 %>% mutate(BAPS4 = as.factor(BAPS4))
data1 <- data1 %>% mutate(BAPS5 = as.factor(BAPS5))
data1 <- data1 %>% mutate(BAPS6 = as.factor(BAPS6))
##########################################--------############################################
#########################################--------#############################################
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest
data2 <- mlst %>%
          select(id2, ST) %>% # select columns of interest
            mutate_all(na_if, "-") %>% # transform any - to NA
              mutate_all(na_if, "?") %>% # transform any ? to NA
                  rename(id = id2) # rename id2 column to id
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data2$st <- ifelse(data2$ST == 5, "ST5", # group major STs separately, and minor STs as Others 
                   ifelse(data2$ST == 31, "ST31", 
                          ifelse(data2$ST == 45, "ST45", 
                                 ifelse(data2$ST == 46, "ST46", 
                                        ifelse(data2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(data2$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data2 <- data2 %>% select(-ST) 
##########################################--------############################################
#########################################--------#############################################
# enter SISTR results
sistr <- read_csv('sistr_output.csv')
# generate the id column 
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select id and cgmlst_ST columns 
data3 <- sistr %>%
            select(id, serovar, cgmlst_ST) %>% # select columns of interest
            mutate_all(na_if, "-") %>% # transform any - to NA
              mutate_all(na_if, "?") # transform any ? to NA
# group serovars as Newport or others
data3$serovar <- ifelse(data3$serovar == "Newport", "Newport",
                            "Other serovars") # group genomes to either Newport or other serovars 
##########################################--------############################################
#########################################--------#############################################
# merge datasets
data4 <- left_join(data1, data2, on = "id") # join datasets based on id
data5 <- left_join(data3, data4, on = "id") # join datasets based on id
# filter data for Newport only
data6 <- data5 %>% filter(serovar == "Newport")
# check for missing values
skim(data6)
# group ST missing values as "Other STs"
data6 <- data6 %>% mutate(st = replace_na(st, "Other STs"))
# check the distribution of serovars to make sure there is only Newport
table(data6$serovar)
##########################################--------############################################
#########################################--------#############################################
# cgMLST diversity 
# drop the STs and cgMLSTs with NAs, group by ST and cgMLST and then calculate Simpson's index
cgmlst_div <- data6 %>%
              drop_na(st, cgmlst_ST) %>% # drop NAs 
              mutate(cgmlst_ST = as.factor(cgmlst_ST)) %>% # transform column to categorical 
              select(st, cgmlst_ST) %>% # select columns of interest 
              group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate the index of diversity
              group_by(st) %>% # group by st 
              summarise(simpson = mean(simpson)) %>% # get the mean of the simpson index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # convert into long format
              mutate(strat = "cgMLST") # create a strat column
##########################################--------############################################
#########################################--------#############################################
# BAPS based diversity (calculated across BAPS levels 1 through 6) 
# BAPS level 1
# drop the STs and BAPS1 with NAs, group by ST and BAPS1 and then calculate Simpson's index
baps1 <- data6 %>%
              select(st, BAPS1) %>% # select columns 
              drop_na(st, BAPS1) %>% # drop NAs 
              group_by(st, BAPS1) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format 
              mutate(strat = "BAPS1") # create a strat column 
# BAPS level 2
# drop the STs and BAPS2 with NAs, group by ST and BAPS2 and then calculate Simpson's index
baps2 <- data6 %>%
              select(st, BAPS2) %>%  # select columns 
              drop_na(st, BAPS2) %>% # drop NAs 
              group_by(st, BAPS2) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS2") # create a strat column 
# BAPS level 3
# drop the STs and BAPS3 with NAs, group by ST and BAPS3 and then calculate Simpson's index
baps3 <- data6 %>%
              select(st, BAPS3) %>% # select columns 
              drop_na(st, BAPS3) %>% # drop NAs 
              group_by(st, BAPS3) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS3") # create a strat column 
# BAPS level 4
# drop the STs and BAPS4 with NAs, group by ST and BAPS4 and then calculate Simpson's index
baps4 <- data6 %>%
              select(st, BAPS4) %>% # select columns 
              drop_na(st, BAPS4) %>% # drop NAs 
              group_by(st, BAPS4) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS4") # create a strat column 
# BAPS level 5
# drop the STs and BAPS5 with NAs, group by ST and BAPS5 and then calculate Simpson's index
baps5 <- data6 %>%
              select(st, BAPS5) %>% # select columns 
              drop_na(st, BAPS5) %>% # drop NAs 
              group_by(st, BAPS5) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS5") # create a strat column 
# BAPS level 6
# drop the STs and BAPS6 with NAs, group by ST and BAPS6 and then calculate Simpson's index
baps6 <- data6 %>%
              select(st, BAPS6) %>%  # select columns 
              drop_na(st, BAPS6) %>% # drop NAs 
              group_by(st, BAPS6) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS6") # create a strat column 
##########################################--------############################################
#########################################--------#############################################
# Concatenate all datasets
data7 <- rbind(cgmlst_div, baps1, baps2, baps3, baps4, baps5, baps6) 
# order the ST column
data7$st <- factor(data7$st, levels=c("ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs"))
##########################################--------############################################
#########################################--------#############################################
# plot Simpson's diversity analysis
# order the strat colunmn 
data7$strat <- factor(data7$strat, levels=c("cgMLST", "BAPS6", "BAPS5", "BAPS4", "BAPS3", "BAPS2", "BAPS1"))
sup_fig_3 <- ggplot(data7, aes(x = strat, y = value)) +  # show strata on x-axis and index values on y-axis
  xlab("") + ylab("Index value") + ylim(0,1) +  # set labels for axis and limit for y-axis
  theme_bw() + # set plot background 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 70, hjust = 1, size = 24)) + # change x-axis text font size, angle, and orientation 
  theme(strip.text.x = element_text(size = 30, colour = "black", angle = 0)) +  # customize figure's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  facet_wrap(~st) +  # generate multi-plots for STs
  coord_flip()  # flip x and y axis
sup_fig_3
```

Supplementary Figure S4. BAPS levels 1-6 distribution of sub-groups across ST lineages. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# enter BAPS1-6 data
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# changing column names 
colnames(baps)[1:7] <- c("id", "BAPS1", "BAPS2", "BAPS3", "BAPS4", "BAPS5", "BAPS6") 
# assign baps data to d1
data1 <- baps
# transform BAPS sub-groups to factor (categorical data)
data1 <- data1 %>% mutate(BAPS1 = as.factor(BAPS1))
data1 <- data1 %>% mutate(BAPS2 = as.factor(BAPS2))
data1 <- data1 %>% mutate(BAPS3 = as.factor(BAPS3))
data1 <- data1 %>% mutate(BAPS4 = as.factor(BAPS4))
data1 <- data1 %>% mutate(BAPS5 = as.factor(BAPS5))
data1 <- data1 %>% mutate(BAPS6 = as.factor(BAPS6))
##########################################--------############################################
#########################################--------#############################################
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data2 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # transform any - to NAs
              mutate_all(na_if, "?") %>% # transform any ? to NAs
                  rename(id = id2) # rename column
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data2$st <- ifelse(data2$ST == 5, "ST5",  # aggregate minor STs as others 
                   ifelse(data2$ST == 31, "ST31", 
                          ifelse(data2$ST == 45, "ST45", 
                                 ifelse(data2$ST == 46, "ST46", 
                                        ifelse(data2$ST == 118, "ST118", 
                                               ifelse(data2$ST == 132, "ST132", 
                          ifelse(data2$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data2 <- data2 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter SISTR results
sistr <- read_csv('sistr_output.csv')
# generate the id column 
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select id and cgmlst_ST columns and rename -/? to NAs
data3 <- sistr %>%
            select(id, serovar, cgmlst_ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # transform any - to NA
              mutate_all(na_if, "?") # transform any ? to NA
# group serovars as Newport or others
data3$serovar <- ifelse(data3$serovar == "Newport", "Newport",
                            "Other serovars")
##########################################--------############################################
#########################################--------#############################################
# merge datasets
data4 <- left_join(data1, data2, on = "id") # join datasets on id
data5 <- left_join(data3, data4, on = "id") # join datasets on id
# filter data for Newport only
data6 <- data5 %>% filter(serovar == "Newport")
# check for missing values
skim(data6)
# group ST missing values as "Other STs"
data6 <- data6 %>% mutate(st = replace_na(st, "Other STs"))
# check the distribution of serovars to make sure there is only Newport
table(data6$serovar)
# select only needed columns (exclude columns labeled as serovar and cgmlst_ST)
data7 <- data6 %>% select(-c(serovar, cgmlst_ST))
# check for missing values again
vis_miss(data7)
# another way to check for missing values
skim(data7)
# yet another way to check for missing values
sum(is.na(data7))
##########################################--------############################################
#########################################--------#############################################
# plot data
# BAPS levels 1-6 distributions
# BAPS level 1
# drop the STs and BAPS1 with NAs, group by ST and BAPS1 and then calculate distribution
baps1 <- data7 %>%
              select(st, BAPS1) %>% # select columns 
              drop_na(st, BAPS1) %>% # drop NAs for selected columns 
              group_by(st, BAPS1) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>%  # sum up values to create a column 
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS1") %>% # create a group column 
              rename(baps = BAPS1) # rename a column 
# BAPS level 2
# drop the STs and BAPS2 with NAs, group by ST and BAPS2 and then calculate distribution
baps2 <- data7 %>%
              select(st, BAPS2) %>% # select columns
              drop_na(st, BAPS2) %>% # drop NAs for selected columns 
              group_by(st, BAPS2) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>% # sum up values to create a column
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS2") %>% # create a group column 
              rename(baps = BAPS2) # rename a column 
# BAPS level 3
# drop the STs and BAPS3 with NAs, group by ST and BAPS3 and then calculate distribution
baps3 <- data7 %>%
              select(st, BAPS3) %>% # select columns
              drop_na(st, BAPS3) %>% # drop NAs for selected columns 
              group_by(st, BAPS3) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>% # sum up values to create a column
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS3") %>% # create a group column 
              rename(baps = BAPS3) # rename a column 
# BAPS level 4
# drop the STs and BAPS4 with NAs, group by ST and BAPS4 and then calculate distribution
baps4 <- data7 %>%
              select(st, BAPS4) %>% # select columns
              drop_na(st, BAPS4) %>% # drop NAs for selected columns 
              group_by(st, BAPS4) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>% # sum up values to create a column
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS4") %>% # create a group column 
              rename(baps = BAPS4) # rename a column 
# BAPS level 5
# drop the STs and BAPS5 with NAs, group by ST and BAPS5 and then calculate distribution
baps5 <- data7 %>%
              select(st, BAPS5) %>% # select columns
              drop_na(st, BAPS5) %>% # drop NAs for selected columns 
              group_by(st, BAPS5) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>% # sum up values to create a column
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS5") %>% # create a group column
              rename(baps = BAPS5) # rename a column 
# BAPS level 6
# drop the STs and BAPS6 with NAs, group by ST and BAPS6 and then calculate distribution
baps6 <- data7 %>%
              select(st, BAPS6) %>% # select columns
              drop_na(st, BAPS6) %>% # drop NAs for selected columns 
              group_by(st, BAPS6) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(total = sum(n)) %>% # sum up values to create a column
              mutate(prop = n/total*100) %>% # calculate proportions 
              mutate(group = "BAPS6") %>% # create a group column
              rename(baps = BAPS6) # rename a column 
##########################################--------############################################
#########################################--------#############################################
# Concatenate all datasets
data8 <- rbind(baps1, baps2, baps3, baps4, baps5, baps6)
# order BAPS levels 
data8$group <- factor(data8$group, levels=c("BAPS1", "BAPS2", "BAPS3",
                                              "BAPS4", "BAPS5", "BAPS6"))
##########################################--------############################################
#########################################--------#############################################
# plot data

# transform baps column to numeric
data8 <- data8 %>% mutate(baps = as.numeric(baps))
# order ST data 
data8$st <- factor(data8$st, levels=c("ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs"))
sup_fig_4 <- ggplot(data8, aes(x = baps, y = prop, fill = st)) +  # show BAPS levels on x-axis and proportion on y-axis
  xlab("BAPS sub-groups (haplotypes)") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  theme_bw() + # set plot background 
  theme(axis.text.x = element_text(angle = 0, size = 25)) + # change x-axis text font size and angle
  theme(axis.title.x = element_text(size = 35, face = "bold")) + # change x-axis title font size and face
  theme(axis.title.y = element_text(size = 35, face = "bold")) + # change y-axis title font size and face
  theme(axis.text.y = element_text(angle = 0, hjust = 1, size = 25)) + # change y-axis text font size, angle, and orientation
  theme(legend.title=element_text(size=35, face = "bold")) + # customize legend title
  theme(legend.text=element_text(size=25)) + # customize legend text
  theme(strip.text.x = element_text(size = 35)) +  # customize figure's title, legend, font
  scale_fill_manual(name = "ST", values=c("orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue",
                                          "black", "coral")) +  # add colors and labels for the scale
  geom_col(position = "dodge") +  # position columns side by side
  facet_wrap(~group, ncol = 2)  # generate multi-plots for BAPS with 2 columns
sup_fig_4
```

Figure 4. Distribution of Resfinder-annotated AMR loci across ST lineages. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data1 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # transform - to NA
              mutate_all(na_if, "?") %>% # transform ? to NA
                  rename(id = id2) # rename column
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data1$st <- ifelse(data1$ST == 5, "ST5", # group minor STs as others 
                   ifelse(data1$ST == 31, "ST31", 
                          ifelse(data1$ST == 45, "ST45", 
                                 ifelse(data1$ST == 46, "ST46", 
                                        ifelse(data1$ST == 118, "ST118", 
                                               ifelse(data1$ST == 132, "ST132", 
                          ifelse(data1$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data1 <- data1 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter the Resfinder loci databases (AMR loci)
abx <- read_csv('sabricate_resfinder_output.csv')
# create the id column 
abx$id <- sapply(strsplit(as.character(abx$'#FILE'),'_'), "[", 1)
# change the name of the GENE column 
abx <- abx %>% mutate(gene = GENE)
# select columns of interest
abx1 <- abx %>% select(id, gene)
##########################################--------############################################
#########################################--------#############################################
# enter SISTR results
sistr <- read_csv('sistr_output.csv')
# generate the id column 
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select id and cgmlst_ST columns and rename -/? to NAs
data3 <- sistr %>%
            select(id, serovar) %>% # select columns
            mutate_all(na_if, "-") %>% # transform - to NA
              mutate_all(na_if, "?") # transform ? to NA
# group serovars as Newport or others
data3$serovar <- ifelse(data3$serovar == "Newport", "Newport",
                            "Other serovars")
##########################################--------############################################
#########################################--------#############################################
# merge datasets
data4 <- left_join(data1, abx1, on = "id") # join datasets based on id
data5 <- left_join(data4, data3, on = "id") # join datasets based on id
# filter data for Newport only
data6 <- data5 %>% filter(serovar == "Newport")
# check for NAs
skim(data6)
# group ST missing values as "Other STs"
data6 <- data6 %>% mutate(st = replace_na(st, "Other STs"))
# check for NAs again
skim(data6)
##########################################--------############################################
#########################################--------#############################################
# calculate the proportion of AMR loci for each ST lineage
# calculations for ST5
d2a <- data6 %>% filter(st == "ST5")
# for ST5, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3a <- d2a %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4a <- d3a %>% 
         group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[8]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions 
d5a <- d4a %>% mutate(st = "ST5")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST31
d2b <- data6 %>% filter(st == "ST31")
# for ST31, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3b <- d2b %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4b <- d3b %>% group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[4]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions 
d5b <- d4b %>% mutate(st = "ST31")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST45
d2c <- data6 %>% filter(st == "ST45")
# for ST45, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3c <- d2c %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4c <- d3c %>% group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[6]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions 
d5c <- d4c %>% mutate(st = "ST45")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST46
d2d <- data6 %>% filter(st == "ST46")
# for ST46, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3d <- d2d %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4d <- d3d %>% group_by(gene) %>%  # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[7]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions 
d5d <- d4d %>% mutate(st = "ST46")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST118
d2e <- data6 %>% filter(st == "ST118")
# for ST118, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3e <- d2e %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4e <- d3e %>% group_by(gene) %>%  # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[2]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions 
d5e <- d4e %>% mutate(st = "ST118")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST132
d2f <- data6 %>% filter(st == "ST132")
# for ST132, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3f <- d2f %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4f <- d3f %>% group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[3]) %>% # get the total counts of st
         mutate(prop = (value/total)*100)  # calculate proportions 
d5f <- d4f %>% mutate(st = "ST132")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST350
d2g <- data6 %>% filter(st == "ST350")
# for ST350, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3g <- d2g %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4g <- d3g %>% group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[5]) %>% # get the total counts of st
         mutate(prop = (value/total)*100)  # calculate proportions
d5g <- d4g %>% mutate(st = "ST350")
##########################################--------############################################
#########################################--------#############################################
# calculations for Other ST lineages
d2h <- data6 %>% filter(st == "Other STs")
# for other STs, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3h <- d2h %>% 
    select(id, gene) %>%  # select columns 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace counts equal to 2 by 1
    filter(count <= 1) # filter counts below or equal to 1 
d4h <- d3h %>% group_by(gene) %>% # group by gene 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[1]) %>% # get the total counts of st
         mutate(prop = (value/total)*100) # calculate proportions
d5h <- d4h %>% mutate(st = "Other STs")
##########################################--------############################################
#########################################--------#############################################
# combined datasets
d6 <- rbind(d5a, d5b, d5c, d5d, d5e, d5f, d5g, d5h)
##########################################--------############################################
#########################################--------#############################################
# export data table containing ST and AMR loci information
abx_newport_st <- d6
write.csv(abx_newport_st,"~/Documents/jove_paper/data/abx_newport_st.csv", row.names = FALSE)
#######################################################
# filter AMR loci with proportion higher than or equal to %
d7 <- d6 %>% filter(prop >= 10)
# order STs 
d7$st <- factor(d7$st, levels=c("ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs"))
# plot Figure 8
# create a vector object with the number of unique genes or loci in the data
colourCount = length(unique(d7$gene))
figure_4 <- ggplot(data = d7, mapping = aes(x = prop, y=gene, fill = gene)) +  # show genes on y-axis and proportion on x-axis 
      xlab("Proportion") + ylab("AMR loci") + xlim(0, 105) +  # set labels for axis and limit for x-axis
      theme_bw() + # set the plot background 
      theme(legend.position = "none") + # remove legend 
      theme(axis.text.y = element_text(size = 12)) + # change font size for y-axis text 
      theme(axis.title.y = element_text(size = 35, face = "bold")) + # customize y-axis title font size and face
      theme(axis.title.x = element_text(size = 30, face = "bold")) + # customize x-axis title font size and face
      theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 20)) + # customize x-axis text font size, angle, and orientation 
      theme(strip.text.x = element_text(size = 30, colour = "black", angle = 0)) +  # customize figure's title, legend, font
      scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Accent"))(colourCount)) +  # add colors for the scale
      geom_bar(stat = "identity") +  # show the bars with y values
      facet_wrap(~ st)  # generate multi-plots for STs
figure_4
```

Figure 5 without legend. Phylogeny-guided mapping of hierarchical genotypes along with ST-differentiating AMR loci.

```{r, include = TRUE}
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data1 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # change - to NA
              mutate_all(na_if, "?") %>% # change ? to NA
                  rename(id = id2) # rename column 
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data1$st <- ifelse(data1$ST == 5, "ST5", # aggregate minor STs as Others 
                   ifelse(data1$ST == 31, "ST31", 
                          ifelse(data1$ST == 45, "ST45", 
                                 ifelse(data1$ST == 46, "ST46", 
                                        ifelse(data1$ST == 118, "ST118", 
                                               ifelse(data1$ST == 132, "ST132", 
                          ifelse(data1$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data1 <- data1 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter the Resfinder loci databases (AMR loci)
abx <- read_csv('sabricate_resfinder_output.csv')
# create the id column 
abx$id <- sapply(strsplit(as.character(abx$'#FILE'),'_'), "[", 1)
# change the name of the GENE column 
abx <- abx %>% mutate(gene = GENE)
# select columns of interest
abx1 <- abx %>% select(id, gene)
##########################################--------############################################
#########################################--------#############################################
# join ST and AMR loci data 
d1 <- left_join(data1, abx1, on = "id")
# check for NAs or missing values
skim(d1)
# group ST missing values as "Other STs"
d1 <- d1 %>% mutate(st = replace_na(st, "Other STs"))
# check for NAs again
skim(d1)
##########################################--------############################################
#########################################--------#############################################
# calculate the proportion of AMR loci for each ST lineage
# calculations for ST5
d2a <- d1 %>% filter(st == "ST5")
# for ST5, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3a <- d2a %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4a <- d3a %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[8]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5a <- d4a %>% mutate(st = "ST5")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST31
d2b <- d1 %>% filter(st == "ST31")
# for ST31, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3b <- d2b %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4b <- d3b %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[4]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5b <- d4b %>% mutate(st = "ST31")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST45
d2c <- d1 %>% filter(st == "ST45")
# for ST45, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3c <- d2c %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4c <- d3c %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[6]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5c <- d4c %>% mutate(st = "ST45")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST46
d2d <- d1 %>% filter(st == "ST46")
# for ST46, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3d <- d2d %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4d <- d3d %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[7]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5d <- d4d %>% mutate(st = "ST46")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST118
d2e <- d1 %>% filter(st == "ST118")
# for ST118, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3e <- d2e %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4e <- d3e %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[2]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5e <- d4e %>% mutate(st = "ST118")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST132
d2f <- d1 %>% filter(st == "ST132")
# for ST132, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3f <- d2f %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4f <- d3f %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[3]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5f <- d4f %>% mutate(st = "ST132")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST350
d2g <- d1 %>% filter(st == "ST350")
# for ST350, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3g <- d2g %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4g <- d3g %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[5]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5g <- d4g %>% mutate(st = "ST350")
##########################################--------############################################
#########################################--------#############################################
# calculations for Other ST lineages
d2h <- d1 %>% filter(st == "Other STs")
# for other STs, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3h <- d2h %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4h <- d3h %>% group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[1]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5h <- d4h %>% mutate(st = "Other STs")
##########################################--------############################################
#########################################--------#############################################
# combined datasets
d6 <- rbind(d5a, d5b, d5c, d5d, d5e, d5f, d5g, d5h)
##########################################--------############################################
#########################################--------#############################################
# filter AMR loci with proportion higher than or equal to %
d7 <- d6 %>% filter(prop >= 10)
# create a list of genes with proportion >= 10%
d8 <- d7 %>% select(gene) %>% unique()
# transform dataframe to list of characters or vector
d9 <- pull(d8, gene)
##########################################--------############################################
#########################################--------#############################################
# spread the abx1 table - from long to wide format
# count gene occurrences
abx2 <- abx1 %>% 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count the number of observations 
    filter(count <= 1) # filter genes with 0 or 1 counts 
# spread from long to wide format
abx3 <- spread(abx2, key = gene, value = count)
# select columns of interest
abx4 <- abx3 %>% select(d9)
# replace all NAs with zeros
abx4[is.na(abx4)] <- 0
# merge with original data containing the hierarchical genotypes
d4 <- left_join(d3, abx4, on = "id")
##########################################--------############################################
#########################################--------#############################################
# change binary values from AMR data locus by locus
d4 <- d4 %>% mutate(`aac(6')-Iaa_1` = factor(ifelse(`aac(6')-Iaa_1` == 1, "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)")))
d4 <- d4 %>% mutate(`mdf(A)_1` = factor(ifelse(`mdf(A)_1` == 1, "mdf(A)_1 (present)", "mdf(A)_1 (absent)")))
d4 <- d4 %>% mutate(aadA2_1 = factor(ifelse(aadA2_1 == 1, "aadA2_1 (present)", "aadA2_1 (absent)")))
d4 <- d4 %>% mutate(`ant(3'')-Ia_1` = factor(ifelse(`ant(3'')-Ia_1` == 1, "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_1` = factor(ifelse(`aph(3')-Ia_1` == 1, "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`blaCARB-2_1` = factor(ifelse(`blaCARB-2_1` == 1, "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)")))
d4 <- d4 %>% mutate(catA1_1 = factor(ifelse(catA1_1 == 1, "catA1_1 (present)", "catA1_1 (absent)")))
d4 <- d4 %>% mutate(sul1_5 = factor(ifelse(sul1_5 == 1, "sul1_5 (present)", "sul1_5 (absent)")))
d4 <- d4 %>% mutate(`tet(A)_6` = factor(ifelse(`tet(A)_6` == 1, "tet(A)_6 (present)", "tet(A)_6 (absent)")))
d4 <- d4 %>% mutate(`tet(B)_2` = factor(ifelse(`tet(B)_2` == 1, "tet(B)_2 (present)", "tet(B)_2 (absent)")))
d4 <- d4 %>% mutate(`aph(3'')-Ib_5` = factor(ifelse(`aph(3'')-Ib_5` == 1, "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_9` = factor(ifelse(`aph(3')-Ia_9` == 1, "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)")))
d4 <- d4 %>% mutate(`aph(6)-Id_1` = factor(ifelse(`aph(6)-Id_1` == 1, "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)")))
d4 <- d4 %>% mutate(`blaCMY-2_1` = factor(ifelse(`blaCMY-2_1` == 1, "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)")))
d4 <- d4 %>% mutate(floR_2 = factor(ifelse(floR_2 == 1, "floR_2  (present)", "floR_2 (absent)")))
d4 <- d4 %>% mutate(sul2_2 = factor(ifelse(sul2_2 == 1, "sul2_2 (present)", "sul2_2 (absent)")))
d4 <- d4 %>% mutate(dfrA1_8 = factor(ifelse(dfrA1_8 == 1, "dfrA1_8 (present)", "dfrA1_8 (absent)")))
d4 <- d4 %>% mutate(`mph(A)_2` = factor(ifelse(`mph(A)_2` == 1, "mph(A)_2 (present)", "mph(A)_2 (absent)")))
d4 <- d4 %>% mutate(qnrA1_1 = factor(ifelse(qnrA1_1 == 1, "qnrA1_1 (present)", "qnrA1_1 (absent)")))
##########################################--------############################################
#########################################--------#############################################
# enter the phylogeny data
tree <- read.tree("newport_phylogeny.tree")
# bring data containing hierarchical genotypes and AMR loci
d5 <- d4
# to plot with the phylogeny using ggtree we need to make the id column into index first 
d6 <- column_to_rownames(d5, var = "id") 
# create the tree
# adjusting the parameters to make the tree visible or however you wish requires some trial and error 
tree_plot <- ggtree(tree, layout = "circular") + xlim(-50, NA)
# plot figure 1 - color scheme for each layer of the plot should be chosen based on the user preferences 
figure_5 <- gheatmap(tree_plot, d6, offset=.0, width=50, colnames = FALSE) +  # visualize the tree with metadata
  scale_fill_manual(values = c("coral", "darkblue",
                               "cornflowerblue", "coral", "purple", "red", "brown", "darkseagreen3", "darkblue", "darkgreen", "yellow",
                               "orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue", "black", "coral",
                               "black", "darkgreen", "purple", "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray"),
                    breaks = c("Newport", "Other serovars",
                               "BAPS1 sub-group 1", "BAPS1 sub-group 2", "BAPS1 sub-group 3", "BAPS1 sub-group 4", 
                               "BAPS1 sub-group 5", "BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 8", 
                               "BAPS1 sub-group 9",
                               "ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs",
                               "cgMLST 1468400426", "cgMLST 1271156802", "cgMLST 3336043520", "cgMLST 88443731", "Other cgMLSTs",
                               "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)",
                               "mdf(A)_1 (present)", "mdf(A)_1 (absent)",
                               "aadA2_1 (present)", "aadA2_1 (absent)",
                               "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)",
                               "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)",
                               "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)",
                               "catA1_1 (present)", "catA1_1 (absent)",
                               "sul1_5 (present)", "sul1_5 (absent)", 
                               "tet(A)_6 (present)", "tet(A)_6 (absent)", 
                               "tet(B)_2 (present)", "tet(B)_2 (absent)",
                               "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)",
                               "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)",
                               "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)",
                               "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)",
                               "floR_2  (present)", "floR_2 (absent)",
                               "sul2_2 (present)", "sul2_2 (absent)",
                               "dfrA1_8 (present)", "dfrA1_8 (absent)",
                               "mph(A)_2 (present)", "mph(A)_2 (absent)", 
                               "qnrA1_1 (present)", "qnrA1_1 (absent)"),
                    name="Serovar -> BAPS1 -> ST -> cgMLST -> AMR loci") +  # add colors and labels for the scale 
  theme(legend.title=element_text(size=24, face = "bold"), 
    legend.text=element_text(size=22)) +
        theme(legend.position = "none")  # customize figure's title, legend, font
figure_5
```
Figure 5 with legend. Phylogeny-guided mapping of hierarchical genotypes along with ST-differentiating AMR loci.

```{r, include = TRUE}
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data1 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # change - to NA
              mutate_all(na_if, "?") %>% # change ? to NA
                  rename(id = id2) # rename column 
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data1$st <- ifelse(data1$ST == 5, "ST5", # aggregate minor STs as Others 
                   ifelse(data1$ST == 31, "ST31", 
                          ifelse(data1$ST == 45, "ST45", 
                                 ifelse(data1$ST == 46, "ST46", 
                                        ifelse(data1$ST == 118, "ST118", 
                                               ifelse(data1$ST == 132, "ST132", 
                          ifelse(data1$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data1 <- data1 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter the Resfinder loci databases (AMR loci)
abx <- read_csv('sabricate_resfinder_output.csv')
# create the id column 
abx$id <- sapply(strsplit(as.character(abx$'#FILE'),'_'), "[", 1)
# change the name of the GENE column 
abx <- abx %>% mutate(gene = GENE)
# select columns of interest
abx1 <- abx %>% select(id, gene)
##########################################--------############################################
#########################################--------#############################################
# join ST and AMR loci data 
d1 <- left_join(data1, abx1, on = "id")
# check for NAs or missing values
skim(d1)
# group ST missing values as "Other STs"
d1 <- d1 %>% mutate(st = replace_na(st, "Other STs"))
# check for NAs again
skim(d1)
##########################################--------############################################
#########################################--------#############################################
# calculate the proportion of AMR loci for each ST lineage
# calculations for ST5
d2a <- d1 %>% filter(st == "ST5")
# for ST5, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3a <- d2a %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4a <- d3a %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[8]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5a <- d4a %>% mutate(st = "ST5")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST31
d2b <- d1 %>% filter(st == "ST31")
# for ST31, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3b <- d2b %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4b <- d3b %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[4]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5b <- d4b %>% mutate(st = "ST31")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST45
d2c <- d1 %>% filter(st == "ST45")
# for ST45, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3c <- d2c %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4c <- d3c %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[6]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5c <- d4c %>% mutate(st = "ST45")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST46
d2d <- d1 %>% filter(st == "ST46")
# for ST46, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3d <- d2d %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4d <- d3d %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[7]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5d <- d4d %>% mutate(st = "ST46")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST118
d2e <- d1 %>% filter(st == "ST118")
# for ST118, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3e <- d2e %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4e <- d3e %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[2]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5e <- d4e %>% mutate(st = "ST118")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST132
d2f <- d1 %>% filter(st == "ST132")
# for ST132, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3f <- d2f %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4f <- d3f %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[3]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5f <- d4f %>% mutate(st = "ST132")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST350
d2g <- d1 %>% filter(st == "ST350")
# for ST350, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3g <- d2g %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4g <- d3g %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[5]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5g <- d4g %>% mutate(st = "ST350")
##########################################--------############################################
#########################################--------#############################################
# calculations for Other ST lineages
d2h <- d1 %>% filter(st == "Other STs")
# for other STs, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3h <- d2h %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4h <- d3h %>% group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[1]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5h <- d4h %>% mutate(st = "Other STs")
##########################################--------############################################
#########################################--------#############################################
# combined datasets
d6 <- rbind(d5a, d5b, d5c, d5d, d5e, d5f, d5g, d5h)
##########################################--------############################################
#########################################--------#############################################
# filter AMR loci with proportion higher than or equal to %
d7 <- d6 %>% filter(prop >= 10)
# create a list of genes with proportion >= 10%
d8 <- d7 %>% select(gene) %>% unique()
# transform dataframe to list of characters or vector
d9 <- pull(d8, gene)
##########################################--------############################################
#########################################--------#############################################
# spread the abx1 table - from long to wide format
# count gene occurrences
abx2 <- abx1 %>% 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count the number of observations 
    filter(count <= 1) # filter genes with 0 or 1 counts 
# spread from long to wide format
abx3 <- spread(abx2, key = gene, value = count)
# select columns of interest
abx4 <- abx3 %>% select(d9)
# replace all NAs with zeros
abx4[is.na(abx4)] <- 0
# merge with original data containing the hierarchical genotypes
d4 <- left_join(d3, abx4, on = "id")
##########################################--------############################################
#########################################--------#############################################
# change binary values from AMR data locus by locus
d4 <- d4 %>% mutate(`aac(6')-Iaa_1` = factor(ifelse(`aac(6')-Iaa_1` == 1, "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)")))
d4 <- d4 %>% mutate(`mdf(A)_1` = factor(ifelse(`mdf(A)_1` == 1, "mdf(A)_1 (present)", "mdf(A)_1 (absent)")))
d4 <- d4 %>% mutate(aadA2_1 = factor(ifelse(aadA2_1 == 1, "aadA2_1 (present)", "aadA2_1 (absent)")))
d4 <- d4 %>% mutate(`ant(3'')-Ia_1` = factor(ifelse(`ant(3'')-Ia_1` == 1, "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_1` = factor(ifelse(`aph(3')-Ia_1` == 1, "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`blaCARB-2_1` = factor(ifelse(`blaCARB-2_1` == 1, "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)")))
d4 <- d4 %>% mutate(catA1_1 = factor(ifelse(catA1_1 == 1, "catA1_1 (present)", "catA1_1 (absent)")))
d4 <- d4 %>% mutate(sul1_5 = factor(ifelse(sul1_5 == 1, "sul1_5 (present)", "sul1_5 (absent)")))
d4 <- d4 %>% mutate(`tet(A)_6` = factor(ifelse(`tet(A)_6` == 1, "tet(A)_6 (present)", "tet(A)_6 (absent)")))
d4 <- d4 %>% mutate(`tet(B)_2` = factor(ifelse(`tet(B)_2` == 1, "tet(B)_2 (present)", "tet(B)_2 (absent)")))
d4 <- d4 %>% mutate(`aph(3'')-Ib_5` = factor(ifelse(`aph(3'')-Ib_5` == 1, "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_9` = factor(ifelse(`aph(3')-Ia_9` == 1, "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)")))
d4 <- d4 %>% mutate(`aph(6)-Id_1` = factor(ifelse(`aph(6)-Id_1` == 1, "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)")))
d4 <- d4 %>% mutate(`blaCMY-2_1` = factor(ifelse(`blaCMY-2_1` == 1, "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)")))
d4 <- d4 %>% mutate(floR_2 = factor(ifelse(floR_2 == 1, "floR_2  (present)", "floR_2 (absent)")))
d4 <- d4 %>% mutate(sul2_2 = factor(ifelse(sul2_2 == 1, "sul2_2 (present)", "sul2_2 (absent)")))
d4 <- d4 %>% mutate(dfrA1_8 = factor(ifelse(dfrA1_8 == 1, "dfrA1_8 (present)", "dfrA1_8 (absent)")))
d4 <- d4 %>% mutate(`mph(A)_2` = factor(ifelse(`mph(A)_2` == 1, "mph(A)_2 (present)", "mph(A)_2 (absent)")))
d4 <- d4 %>% mutate(qnrA1_1 = factor(ifelse(qnrA1_1 == 1, "qnrA1_1 (present)", "qnrA1_1 (absent)")))
##########################################--------############################################
#########################################--------#############################################
# enter the phylogeny data
tree <- read.tree("newport_phylogeny.tree")
# bring data containing hierarchical genotypes and AMR loci
d5 <- d4
# to plot with the phylogeny using ggtree we need to make the id column into index first 
d6 <- column_to_rownames(d5, var = "id") 
# create the tree
# adjusting the parameters to make the tree visible or however you wish requires some trial and error 
tree_plot <- ggtree(tree, layout = "circular") + xlim(-50, NA)
# plot figure 1 - color scheme for each layer of the plot should be chosen based on the user preferences 
figure_5b <- gheatmap(tree_plot, d6, offset=.0, width=50, colnames = FALSE) +  # visualize the tree with metadata
  scale_fill_manual(values = c("coral", "darkblue",
                               "cornflowerblue", "coral", "purple", "red", "brown", "darkseagreen3", "darkblue", "darkgreen", "yellow",
                               "orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue", "black", "coral",
                               "black", "darkgreen", "purple", "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray"),
                    breaks = c("Newport", "Other serovars",
                               "BAPS1 sub-group 1", "BAPS1 sub-group 2", "BAPS1 sub-group 3", "BAPS1 sub-group 4", 
                               "BAPS1 sub-group 5", "BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 8", 
                               "BAPS1 sub-group 9",
                               "ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs",
                               "cgMLST 1468400426", "cgMLST 1271156802", "cgMLST 3336043520", "cgMLST 88443731", "Other cgMLSTs",
                               "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)",
                               "mdf(A)_1 (present)", "mdf(A)_1 (absent)",
                               "aadA2_1 (present)", "aadA2_1 (absent)",
                               "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)",
                               "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)",
                               "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)",
                               "catA1_1 (present)", "catA1_1 (absent)",
                               "sul1_5 (present)", "sul1_5 (absent)", 
                               "tet(A)_6 (present)", "tet(A)_6 (absent)", 
                               "tet(B)_2 (present)", "tet(B)_2 (absent)",
                               "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)",
                               "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)",
                               "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)",
                               "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)",
                               "floR_2  (present)", "floR_2 (absent)",
                               "sul2_2 (present)", "sul2_2 (absent)",
                               "dfrA1_8 (present)", "dfrA1_8 (absent)",
                               "mph(A)_2 (present)", "mph(A)_2 (absent)", 
                               "qnrA1_1 (present)", "qnrA1_1 (absent)"),
                    name="Serovar -> BAPS1 -> ST -> cgMLST -> AMR loci") +  # add colors and labels for the scale 
  theme(legend.title=element_text(size=24, face = "bold"), 
    legend.text=element_text(size=22)) # customize figure's title, legend, font
figure_5b
```
