---
title: "Figure 2"
author: "Joao Carlos Gomes-Neto"
date: "8/2/2021"
output: html_document
---

When beginning running this script, you can remove the # that comes before the install.packages() function, but when running it the second time around, and subsequently, comment the function out using #. That way you avoid creating issues with dependencies, and different versions of the program. 

```{r, include = TRUE}
# install tidyverse 
# install.packages("tidyverse")
# install skimr 
# install.packages("skimr")
# install vegan 
# install.packages("vegan")
# install forcats
# install.packages("forcats")
# install naniar
# install.packages("naniar")
# install ggpubr
# install.packages("ggpubr")
# install ggrepel
# install.packages("ggrepel")
# install reshape2
# install.packages("reshape2")
# install reshape2
# install.packages("RColorBrewer")
# install ggtree 
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ggtree")
# installation of ggtree will prompt a question about installation - answer is "a" to install/update all dependencies 
```

How to activate packages prior to utilization.

```{r, include = TRUE}
# Load previously installed packages
library(tidyverse)
library(skimr)
library(vegan)
library(forcats)
library(naniar)
library(ggtree)
library(ggpubr)
library(ggrepel)
library(reshape2)
library(RColorBrewer)
```

Enter and quality control all genotypic data including: serovar-predictions (generated by SISTR), BAPS level 1 (generated by fastbaps), ST lineages (generated by mlst), 
and cgMLST variants (generated by SISTR). All input files were generated by the describe programs which are part of the computational platform called ProkEvo.

```{r, include = TRUE}
# enter the BAPS data 
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# check the first six rows of the dataset
head(baps)
# changing the first two column names because for hierarchical analysis we only use BAPS1
colnames(baps)[1:2] <- c("id", "baps1") 
# check the first six rows of the datset again to see the change in column names
head(baps)
# select columns id and baps_1
baps1 <- baps %>% 
              select(id, baps1)
# check the first six rows of the dataset
head(baps1)
# quality control baps1 data 
skim(baps1)
# using a plotting strategy to check for missing values
vis_miss(baps1)
# rename BAPS level 1 sub-groups or haplotypes 
baps1$baps_1 <- ifelse(baps1$baps1 == 1, "BAPS1 sub-group 1", # the ifelse functions tests for different conditions prior to assigning groups to one category or another 
                   ifelse(baps1$baps1 == 2, "BAPS1 sub-group 2", 
                          ifelse(baps1$baps1 == 3, "BAPS1 sub-group 3", 
                                 ifelse(baps1$baps1 == 4, "BAPS1 sub-group 4", 
                                        ifelse(baps1$baps1 == 5, "BAPS1 sub-group 5", 
                                               ifelse(baps1$baps1 == 6, "BAPS1 sub-group 6", 
                                                       ifelse(baps1$baps1 == 7, "BAPS1 sub-group 7", 
                                                               ifelse(baps1$baps1 == 8, "BAPS1 sub-group 8", "BAPS1 sub-group 9"))))))))
##########################################--------############################################
#########################################--------#############################################
# enter MLST results
mlst <- read_csv('salmonellast_output.csv')
# check the first six rows of the dataset
head(mlst)
# generate the id column by deriving it from the FILE column
mlst$id <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# check the first six rows of the dataset
head(mlst)
# select the id and ST columns 
mlst1 <- mlst %>%
          select(id, ST) # select id and ST columns 
# use the table() function to detect extraneous characters such as "-" or "?" in the data - those are ST misclassifications
# check for the presence of ST misclassification in the ST column 
table(mlst1$ST)
# check for missing values
skim(mlst1)
# mutate the "-" character to NA (missing value)
mlst1 <- mlst1 %>%
            mutate_all(na_if, "-") # fill all - with NAs
# check the first six rows of the dataset
head(mlst1)
# quality control baps1 data 
skim(mlst1)
# using a plotting strategy to check for missing values
vis_miss(mlst1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# Enter SISTR results
sistr <- read_csv('sistr_output.csv')
# check the first six rows of the dataset
head(sistr)
# generate the id column by deriving it from the genome column
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select the id and cgmlst_ST columns 
sistr1 <- sistr %>%
            select(id, serovar_cgmlst, cgmlst_ST) # select id, serovar_cgmlst, and cgmlst_ST columns 
# check the first six rows of the dataset
head(sistr1)
# quality control baps1 data 
skim(sistr1)
# using a plotting strategy to check for missing values
vis_miss(sistr1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# combine baps1, mlst1, and sistr1 datasets
d1 <- left_join(baps1, mlst1, on = "id") # merge datasets on identical ids 
d2 <- left_join(d1, sistr1, on = "id") # merge datasets on identical ids 
# check data dimensionality 
dim(d2)
# check the first six rows of the dataset
head(d2)
# quality control baps1 data 
skim(d2)
# using a plotting strategy to check for missing values
vis_miss(d2)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# group all genomes not classified as Newport as "Other serovars"
# create a new column called serovar that contains the binary classification to group all SISTR-based misclassified genomes
d2$serovar <- ifelse(d2$serovar_cgmlst == "Newport", "Newport",
                            "Other serovars") # classify serovar_cgmlst into Newport or Others 
# check the first six rows of the dataset
head(d2)
# check the composition of the serovar column 
table(d2$serovar)
##########################################--------############################################
#########################################--------#############################################
# no transformation is needed for the baps1 dataset - no groupings or aggregations are needed 
##########################################--------############################################
#########################################--------#############################################
# check ST distribution to focus on major STs for all subsequent analyses by grouping by ST and counting
st_dist <- d2 %>%
  group_by(ST) %>% # group by the ST column
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange the counts in decreasing order 
st_dist
# based on the frequency analysis, the following STs were not aggregrated: ST118, ST45, ST5, ST132, ST31, ST350, and ST46
# create a new st column
d2$st <- ifelse(d2$ST == 5, "ST5", # create a new ST column for which minor STs are aggregated as Others 
                   ifelse(d2$ST == 31, "ST31", 
                          ifelse(d2$ST == 45, "ST45", 
                                 ifelse(d2$ST == 46, "ST46", 
                                        ifelse(d2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(d2$ST == 350, "ST350", "Other STs")))))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# check cgMLST distribution by ST to focus on major cgMLSTs for all subsequent analyses by grouping by STs and cgMLSTs and counting
cgmlst_dist <- d2 %>%
  group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange in descending order 
cgmlst_dist
# for the purposes of this analysis only the top four most frequent cgMLSTs were selected and respectively renamed
d2 <- mutate(d2, cgmlst = ifelse(cgmlst_ST %in% 1468400426, "cgMLST 1468400426", # create a new cgMLST column while aggregating minor cgMLST variants 
                                     ifelse(cgmlst_ST %in% 88443731, "cgMLST 88443731", 
                                            ifelse(cgmlst_ST %in% 1271156802, "cgMLST 1271156802",
                                                 ifelse(cgmlst_ST %in% 3336043520, "cgMLST 3336043520",  
                                            "Other cgMLSTs")))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# select only needed columns for all subsequent analyses 
d3 <- d2 %>% 
          select(id, serovar, baps_1, st, cgmlst) # select columns of interest which are described within parenthesis 
# check data dimensionality to make sure it matches that of d2 
dim(d3)
# check the first six rows of the dataset
head(d3)
# quality control baps1 data 
skim(d3)
# using a plotting strategy to check for missing values
vis_miss(d3)
# make sure there is no NA in the dataset
sum(is.na(d3))
# replace NA in the ST column
d3 <- d3 %>% mutate(st = replace_na(st, "Other STs")) 
# make sure there is no NA in the dataset
sum(is.na(d3))
# check the distribution of serovars, baps1, st, and cgmlst classifications
table(d3$serovar)
table(d3$baps_1)
table(d3$st)
table(d3$cgmlst)
# check the first six rows in the dataset 
head(d3)
# check the last six rows in the dataset 
tail(d3)
# check for missing values
vis_miss(d3)
# check the data characteristics 
str(d3)
```

Figure 2. Visualize the frequency-based distribution of hierarchical genotypes. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# bring back the metadata file
d5 <- d3 
##########################################--------############################################
#########################################--------#############################################
# plot Serovar distribution
# calculate the relative frequencies for serovar
# drop the serovars with NAs, group by serovar and then calculate the frequencies
serovar_data <- d5 %>%
              drop_na(serovar) %>% # remove NA based on serovar
              select(serovar) %>% # select the serovar column
              group_by(serovar) %>% # group_by serovar
              summarise(n = n()) %>% # count the number of genomes per serovar classification
              mutate(prop_serovar = n/sum(n)*100) # calculate the proportion of each serovar
# order serovar groups 
serovar_data$serovar <- factor(serovar_data$serovar, levels=c("Other serovars", "Newport"))
# plot the data 
sero_plot <- ggplot(serovar_data, aes(x = prop_serovar, y = serovar)) +  # show serovars on y-axis and proportion on x-axis
  ylab("") + xlab("Proportion") + xlim(0, 100) +  # set labels for axis and limit for x-axis
  ggtitle("A") +  # add title for the plot
  theme_bw() + # choose the plot background
  theme(legend.position = "none") + # remove legend
  theme(axis.text.y = element_text(size = 30)) + # set the font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # set the font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # set the font size for x-axis title
  theme(axis.text.x = element_text(angle = 0, hjust = 0.7, size = 26)) + # set the font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue")  # fill the bars that represent the values with blue
##########################################--------############################################
#########################################--------#############################################
# plot BAPS1 sub-group distribution for Newport data only
# calculate the relative frequencies for serovar and BAPS1 sub-group
# drop the serovars and BAPS1 sub-groups records with NAs, group by serovar and BAPS1 sub-group and then calculate the frequencies
baps_data <- d5 %>%
              drop_na(serovar, baps_1) %>% # drop NAs for serovar and baps_1 columns
              select(serovar, baps_1) %>% # select serovar and baps_1 columns
              group_by(serovar, baps_1) %>% # group data based on serovar and baps_1
              summarise(n = n()) %>% # count the number of observations per group
              mutate(prop = n/sum(n)*100) %>% # calculate the propotion for each group
              filter(serovar != "Other serovars") # filter out genomes classified as Other serovars 
# re-order the baps_1 column
baps_data$baps_1 <- factor(baps_data$baps_1, levels=c("BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 4", "BAPS1 sub-group 1", "BAPS1 sub-group 8"))
# plot the data 
baps_plot <- ggplot(baps_data, aes(x = baps_1, y = prop)) +  # show BAPS sub-groups on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 105) +  # set labels for axis and limit for y-axis
  ggtitle("B") +  # add title for the plot
  theme_bw() + # setting plot background
  theme(legend.position = "none") + # remove legend 
  theme(axis.text.y = element_text(size = 30)) + # change font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change font size for x-axis title 
  theme(axis.text.x = element_text(angle = 0, size = 30)) + # change font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font 
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# plot ST distribution
# calculate the relative frequencies for serovar and ST
# drop the serovars and STs with NAs, group by serovar and ST and then calculate the frequencies
st_data <- d5 %>%
              drop_na(serovar, st) %>% # drop NAs for serovar and st columns 
              select(serovar, st) %>% # select serovar and st columns 
              group_by(serovar, st) %>% # group data based on serovar and st columns 
              summarise(n = n()) %>% # count the number of observations 
              mutate(prop_st = n/sum(n)*100) %>% # calculate the proportion by groups 
              filter(serovar != "Other serovars") # filter out data for Other serovars 

# re-order the st column
st_data$st <- factor(st_data$st, levels=c("Other STs", "ST46", "ST350", "ST31", "ST132", "ST5", "ST45",
                                          "ST118"))
# plot the data
st_plot <- ggplot(st_data, aes(x = st, y = prop_st)) +  # show STs on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 105) +  # set labels for axis and limit for y-axis
  ggtitle("C") +  # add title for the plot
  theme_bw() + # set the plot background 
  theme(legend.position = "none") + # remove legends 
  theme(axis.text.y = element_text(size = 28)) + # change font size for y-axis text 
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change font size for y-axis title
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change font size for x-axis title 
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 28)) + # change font size for x-axis text 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# plot cgMLST distribution
# calculate the relative frequencies for serovar and cgMLST
# drop the serovars and cgMLSTs with NAs, group by serovar and cgMLST and then calculate the frequencies
cgmlst_data <- d5 %>%
              drop_na(serovar, cgmlst) %>% # drop NAs for serovar and cgmlst columns 
              group_by(serovar, cgmlst) %>% # group data based on serovar and cgmlst columns 
              summarise(n = n()) %>% # count the number of observations 
              mutate(prop_cgmlst = n/sum(n)*100) %>% # calculate the proportions 
              filter(serovar != "Other serovars") # filter out data for Other serovars 

# re-order cgmlst column
cgmlst_data$cgmlst <- factor(cgmlst_data$cgmlst, levels=c("Other cgMLSTs", "cgMLST 88443731", "cgMLST 3336043520", "cgMLST 1271156802", "cgMLST 1468400426"))
# plot the data 
cgmlst_plot <- ggplot(cgmlst_data, aes(x = cgmlst, y = prop_cgmlst)) +  # show cgMLSTs on x-axis and proportion on y-axis
  xlab("") + ylab("Proportion") + ylim(0, 100) +  # set labels for axis and limit for y-axis
  ggtitle("D") +  # add title for the plot
  theme_bw() + # set plot background 
  theme(legend.position = "none") + # remove legend 
  theme(axis.text.y = element_text(size = 28)) + # set the font size for y-axis text 
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # set the font size for x-axis title
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size = 28)) + # set the font size for x-axis text, orientations, and angle 
  theme(plot.title = element_text(size = 40, face = "bold")) +  # customize plot's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  coord_flip()  # flip x and y axis
##########################################--------############################################
#########################################--------#############################################
# combine all plots in one to make Figure 2 based on 2 columns x 2 rows 
figure_2  <- ggarrange(sero_plot, baps_plot, st_plot, cgmlst_plot,
                  nrow = 2, ncol = 2, widths = c(10, 10, 10, 10),
                heights = c(1, 1, 1, 1))
figure_2
```

