---
title: "Figure 5"
author: "Joao Carlos Gomes-Neto"
date: "8/2/2021"
output: html_document
---

When beginning running this script, you can remove the # that comes before the install.packages() function, but when running it the second time around, and subsequently, comment the function out using #. That way you avoid creating issues with dependencies, and different versions of the program. 

```{r, include = TRUE}
# install tidyverse 
# install.packages("tidyverse")
# install skimr 
# install.packages("skimr")
# install vegan 
# install.packages("vegan")
# install forcats
# install.packages("forcats")
# install naniar
# install.packages("naniar")
# install ggpubr
# install.packages("ggpubr")
# install ggrepel
# install.packages("ggrepel")
# install reshape2
# install.packages("reshape2")
# install reshape2
# install.packages("RColorBrewer")
# install ggtree 
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ggtree")
# installation of ggtree will prompt a question about installation - answer is "a" to install/update all dependencies 
```

How to activate packages prior to utilization.

```{r, include = TRUE}
# Load previously installed packages
library(tidyverse)
library(skimr)
library(vegan)
library(forcats)
library(naniar)
library(ggtree)
library(ggpubr)
library(ggrepel)
library(reshape2)
library(RColorBrewer)
```

Enter and quality control all genotypic data including: serovar-predictions (generated by SISTR), BAPS level 1 (generated by fastbaps), ST lineages (generated by mlst), 
and cgMLST variants (generated by SISTR). All input files were generated by the describe programs which are part of the computational platform called ProkEvo.

```{r, include = TRUE}
# enter the BAPS data 
baps <- read_csv('~/Documents/jove_paper/data/fastbaps_partition_baps_prior_l6.csv')
# check the first six rows of the dataset
head(baps)
# changing the first two column names because for hierarchical analysis we only use BAPS1
colnames(baps)[1:2] <- c("id", "baps1") 
# check the first six rows of the datset again to see the change in column names
head(baps)
# select columns id and baps_1
baps1 <- baps %>% 
              select(id, baps1)
# check the first six rows of the dataset
head(baps1)
# quality control baps1 data 
skim(baps1)
# using a plotting strategy to check for missing values
vis_miss(baps1)
# rename BAPS level 1 sub-groups or haplotypes 
baps1$baps_1 <- ifelse(baps1$baps1 == 1, "BAPS1 sub-group 1", # the ifelse functions tests for different conditions prior to assigning groups to one category or another 
                   ifelse(baps1$baps1 == 2, "BAPS1 sub-group 2", 
                          ifelse(baps1$baps1 == 3, "BAPS1 sub-group 3", 
                                 ifelse(baps1$baps1 == 4, "BAPS1 sub-group 4", 
                                        ifelse(baps1$baps1 == 5, "BAPS1 sub-group 5", 
                                               ifelse(baps1$baps1 == 6, "BAPS1 sub-group 6", 
                                                       ifelse(baps1$baps1 == 7, "BAPS1 sub-group 7", 
                                                               ifelse(baps1$baps1 == 8, "BAPS1 sub-group 8", "BAPS1 sub-group 9"))))))))
##########################################--------############################################
#########################################--------#############################################
# enter MLST results
mlst <- read_csv('~/Documents/jove_paper/data/salmonellast_output.csv')
# check the first six rows of the dataset
head(mlst)
# generate the id column by deriving it from the FILE column
mlst$id <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# check the first six rows of the dataset
head(mlst)
# select the id and ST columns 
mlst1 <- mlst %>%
          select(id, ST) # select id and ST columns 
# use the table() function to detect extraneous characters such as "-" or "?" in the data - those are ST misclassifications
# check for the presence of ST misclassification in the ST column 
table(mlst1$ST)
# check for missing values
skim(mlst1)
# mutate the "-" character to NA (missing value)
mlst1 <- mlst1 %>%
            mutate_all(na_if, "-") # fill all - with NAs
# check the first six rows of the dataset
head(mlst1)
# quality control baps1 data 
skim(mlst1)
# using a plotting strategy to check for missing values
vis_miss(mlst1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# Enter SISTR results
sistr <- read_csv('~/Documents/jove_paper/data/sistr_output.csv')
# check the first six rows of the dataset
head(sistr)
# generate the id column by deriving it from the genome column
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select the id and cgmlst_ST columns 
sistr1 <- sistr %>%
            select(id, serovar_cgmlst, cgmlst_ST) # select id, serovar_cgmlst, and cgmlst_ST columns 
# check the first six rows of the dataset
head(sistr1)
# quality control baps1 data 
skim(sistr1)
# using a plotting strategy to check for missing values
vis_miss(sistr1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# combine baps1, mlst1, and sistr1 datasets
d1 <- left_join(baps1, mlst1, on = "id") # merge datasets on identical ids 
d2 <- left_join(d1, sistr1, on = "id") # merge datasets on identical ids 
# check data dimensionality 
dim(d2)
# check the first six rows of the dataset
head(d2)
# quality control baps1 data 
skim(d2)
# using a plotting strategy to check for missing values
vis_miss(d2)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# group all genomes not classified as Newport as "Other serovars"
# create a new column called serovar that contains the binary classification to group all SISTR-based misclassified genomes
d2$serovar <- ifelse(d2$serovar_cgmlst == "Newport", "Newport",
                            "Other serovars") # classify serovar_cgmlst into Newport or Others 
# check the first six rows of the dataset
head(d2)
# check the composition of the serovar column 
table(d2$serovar)
##########################################--------############################################
#########################################--------#############################################
# no transformation is needed for the baps1 dataset - no groupings or aggregations are needed 
##########################################--------############################################
#########################################--------#############################################
# check ST distribution to focus on major STs for all subsequent analyses by grouping by ST and counting
st_dist <- d2 %>%
  group_by(ST) %>% # group by the ST column
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange the counts in decreasing order 
st_dist
# based on the frequency analysis, the following STs were not aggregrated: ST118, ST45, ST5, ST132, ST31, ST350, and ST46
# create a new st column
d2$st <- ifelse(d2$ST == 5, "ST5", # create a new ST column for which minor STs are aggregated as Others 
                   ifelse(d2$ST == 31, "ST31", 
                          ifelse(d2$ST == 45, "ST45", 
                                 ifelse(d2$ST == 46, "ST46", 
                                        ifelse(d2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(d2$ST == 350, "ST350", "Other STs")))))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# check cgMLST distribution by ST to focus on major cgMLSTs for all subsequent analyses by grouping by STs and cgMLSTs and counting
cgmlst_dist <- d2 %>%
  group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange in descending order 
cgmlst_dist
# for the purposes of this analysis only the top four most frequent cgMLSTs were selected and respectively renamed
d2 <- mutate(d2, cgmlst = ifelse(cgmlst_ST %in% 1468400426, "cgMLST 1468400426", # create a new cgMLST column while aggregating minor cgMLST variants 
                                     ifelse(cgmlst_ST %in% 88443731, "cgMLST 88443731", 
                                            ifelse(cgmlst_ST %in% 1271156802, "cgMLST 1271156802",
                                                 ifelse(cgmlst_ST %in% 3336043520, "cgMLST 3336043520",  
                                            "Other cgMLSTs")))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# select only needed columns for all subsequent analyses 
d3 <- d2 %>% 
          select(id, serovar, baps_1, st, cgmlst) # select columns of interest which are described within parenthesis 
# check data dimensionality to make sure it matches that of d2 
dim(d3)
# check the first six rows of the dataset
head(d3)
# quality control baps1 data 
skim(d3)
# using a plotting strategy to check for missing values
vis_miss(d3)
# make sure there is no NA in the dataset
sum(is.na(d3))
# replace NA in the ST column
d3 <- d3 %>% mutate(st = replace_na(st, "Other STs")) 
# make sure there is no NA in the dataset
sum(is.na(d3))
# check the distribution of serovars, baps1, st, and cgmlst classifications
table(d3$serovar)
table(d3$baps_1)
table(d3$st)
table(d3$cgmlst)
# check the first six rows in the dataset 
head(d3)
# check the last six rows in the dataset 
tail(d3)
# check for missing values
vis_miss(d3)
# check the data characteristics 
str(d3)
```

Figure 5 without legend. Phylogeny-guided mapping of hierarchical genotypes along with ST-differentiating AMR loci.

```{r, include = TRUE}
# enter the ST data 
mlst <- read_csv('~/Documents/jove_paper/data/salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data1 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # change - to NA
              mutate_all(na_if, "?") %>% # change ? to NA
                  rename(id = id2) # rename column 
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data1$st <- ifelse(data1$ST == 5, "ST5", # aggregate minor STs as Others 
                   ifelse(data1$ST == 31, "ST31", 
                          ifelse(data1$ST == 45, "ST45", 
                                 ifelse(data1$ST == 46, "ST46", 
                                        ifelse(data1$ST == 118, "ST118", 
                                               ifelse(data1$ST == 132, "ST132", 
                          ifelse(data1$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data1 <- data1 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter the Resfinder loci databases (AMR loci)
abx <- read_csv('~/Documents/jove_paper/data/sabricate_resfinder_output.csv')
# create the id column 
abx$id <- sapply(strsplit(as.character(abx$'#FILE'),'_'), "[", 1)
# change the name of the GENE column 
abx <- abx %>% mutate(gene = GENE)
# select columns of interest
abx1 <- abx %>% select(id, gene)
##########################################--------############################################
#########################################--------#############################################
# join ST and AMR loci data 
d1 <- left_join(data1, abx1, on = "id")
# check for NAs or missing values
skim(d1)
# group ST missing values as "Other STs"
d1 <- d1 %>% mutate(st = replace_na(st, "Other STs"))
# check for NAs again
skim(d1)
##########################################--------############################################
#########################################--------#############################################
# calculate the proportion of AMR loci for each ST lineage
# calculations for ST5
d2a <- d1 %>% filter(st == "ST5")
# for ST5, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3a <- d2a %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4a <- d3a %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[8]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5a <- d4a %>% mutate(st = "ST5")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST31
d2b <- d1 %>% filter(st == "ST31")
# for ST31, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3b <- d2b %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4b <- d3b %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[4]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5b <- d4b %>% mutate(st = "ST31")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST45
d2c <- d1 %>% filter(st == "ST45")
# for ST45, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3c <- d2c %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4c <- d3c %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[6]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5c <- d4c %>% mutate(st = "ST45")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST46
d2d <- d1 %>% filter(st == "ST46")
# for ST46, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3d <- d2d %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4d <- d3d %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[7]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5d <- d4d %>% mutate(st = "ST46")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST118
d2e <- d1 %>% filter(st == "ST118")
# for ST118, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3e <- d2e %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4e <- d3e %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[2]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5e <- d4e %>% mutate(st = "ST118")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST132
d2f <- d1 %>% filter(st == "ST132")
# for ST132, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3f <- d2f %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4f <- d3f %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[3]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5f <- d4f %>% mutate(st = "ST132")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST350
d2g <- d1 %>% filter(st == "ST350")
# for ST350, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3g <- d2g %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4g <- d3g %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[5]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5g <- d4g %>% mutate(st = "ST350")
##########################################--------############################################
#########################################--------#############################################
# calculations for Other ST lineages
d2h <- d1 %>% filter(st == "Other STs")
# for other STs, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3h <- d2h %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4h <- d3h %>% group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[1]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5h <- d4h %>% mutate(st = "Other STs")
##########################################--------############################################
#########################################--------#############################################
# combined datasets
d6 <- rbind(d5a, d5b, d5c, d5d, d5e, d5f, d5g, d5h)
##########################################--------############################################
#########################################--------#############################################
# filter AMR loci with proportion higher than or equal to %
d7 <- d6 %>% filter(prop >= 10)
# create a list of genes with proportion >= 10%
d8 <- d7 %>% select(gene) %>% unique()
# transform dataframe to list of characters or vector
d9 <- pull(d8, gene)
##########################################--------############################################
#########################################--------#############################################
# spread the abx1 table - from long to wide format
# count gene occurrences
abx2 <- abx1 %>% 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count the number of observations 
    filter(count <= 1) # filter genes with 0 or 1 counts 
# spread from long to wide format
abx3 <- spread(abx2, key = gene, value = count)
# select columns of interest
abx4 <- abx3 %>% select(d9)
# replace all NAs with zeros
abx4[is.na(abx4)] <- 0
# merge with original data containing the hierarchical genotypes
d4 <- left_join(d3, abx4, on = "id")
##########################################--------############################################
#########################################--------#############################################
# change binary values from AMR data locus by locus
d4 <- d4 %>% mutate(`aac(6')-Iaa_1` = factor(ifelse(`aac(6')-Iaa_1` == 1, "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)")))
d4 <- d4 %>% mutate(`mdf(A)_1` = factor(ifelse(`mdf(A)_1` == 1, "mdf(A)_1 (present)", "mdf(A)_1 (absent)")))
d4 <- d4 %>% mutate(aadA2_1 = factor(ifelse(aadA2_1 == 1, "aadA2_1 (present)", "aadA2_1 (absent)")))
d4 <- d4 %>% mutate(`ant(3'')-Ia_1` = factor(ifelse(`ant(3'')-Ia_1` == 1, "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_1` = factor(ifelse(`aph(3')-Ia_1` == 1, "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`blaCARB-2_1` = factor(ifelse(`blaCARB-2_1` == 1, "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)")))
d4 <- d4 %>% mutate(catA1_1 = factor(ifelse(catA1_1 == 1, "catA1_1 (present)", "catA1_1 (absent)")))
d4 <- d4 %>% mutate(sul1_5 = factor(ifelse(sul1_5 == 1, "sul1_5 (present)", "sul1_5 (absent)")))
d4 <- d4 %>% mutate(`tet(A)_6` = factor(ifelse(`tet(A)_6` == 1, "tet(A)_6 (present)", "tet(A)_6 (absent)")))
d4 <- d4 %>% mutate(`tet(B)_2` = factor(ifelse(`tet(B)_2` == 1, "tet(B)_2 (present)", "tet(B)_2 (absent)")))
d4 <- d4 %>% mutate(`aph(3'')-Ib_5` = factor(ifelse(`aph(3'')-Ib_5` == 1, "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_9` = factor(ifelse(`aph(3')-Ia_9` == 1, "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)")))
d4 <- d4 %>% mutate(`aph(6)-Id_1` = factor(ifelse(`aph(6)-Id_1` == 1, "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)")))
d4 <- d4 %>% mutate(`blaCMY-2_1` = factor(ifelse(`blaCMY-2_1` == 1, "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)")))
d4 <- d4 %>% mutate(floR_2 = factor(ifelse(floR_2 == 1, "floR_2  (present)", "floR_2 (absent)")))
d4 <- d4 %>% mutate(sul2_2 = factor(ifelse(sul2_2 == 1, "sul2_2 (present)", "sul2_2 (absent)")))
d4 <- d4 %>% mutate(dfrA1_8 = factor(ifelse(dfrA1_8 == 1, "dfrA1_8 (present)", "dfrA1_8 (absent)")))
d4 <- d4 %>% mutate(`mph(A)_2` = factor(ifelse(`mph(A)_2` == 1, "mph(A)_2 (present)", "mph(A)_2 (absent)")))
d4 <- d4 %>% mutate(qnrA1_1 = factor(ifelse(qnrA1_1 == 1, "qnrA1_1 (present)", "qnrA1_1 (absent)")))
##########################################--------############################################
#########################################--------#############################################
# enter the phylogeny data
tree <- read.tree("~/Documents/jove_paper/data/newport_phylogeny.tree")
# bring data containing hierarchical genotypes and AMR loci
d5 <- d4
# to plot with the phylogeny using ggtree we need to make the id column into index first 
d6 <- column_to_rownames(d5, var = "id") 
# create the tree
# adjusting the parameters to make the tree visible or however you wish requires some trial and error 
tree_plot <- ggtree(tree, layout = "circular") + xlim(-50, NA)
# plot figure 1 - color scheme for each layer of the plot should be chosen based on the user preferences 
figure_5 <- gheatmap(tree_plot, d6, offset=.0, width=50, colnames = FALSE) +  # visualize the tree with metadata
  scale_fill_manual(values = c("coral", "darkblue",
                               "cornflowerblue", "coral", "purple", "red", "brown", "darkseagreen3", "darkblue", "darkgreen", "yellow",
                               "orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue", "black", "coral",
                               "black", "darkgreen", "purple", "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray"),
                    breaks = c("Newport", "Other serovars",
                               "BAPS1 sub-group 1", "BAPS1 sub-group 2", "BAPS1 sub-group 3", "BAPS1 sub-group 4", 
                               "BAPS1 sub-group 5", "BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 8", 
                               "BAPS1 sub-group 9",
                               "ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs",
                               "cgMLST 1468400426", "cgMLST 1271156802", "cgMLST 3336043520", "cgMLST 88443731", "Other cgMLSTs",
                               "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)",
                               "mdf(A)_1 (present)", "mdf(A)_1 (absent)",
                               "aadA2_1 (present)", "aadA2_1 (absent)",
                               "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)",
                               "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)",
                               "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)",
                               "catA1_1 (present)", "catA1_1 (absent)",
                               "sul1_5 (present)", "sul1_5 (absent)", 
                               "tet(A)_6 (present)", "tet(A)_6 (absent)", 
                               "tet(B)_2 (present)", "tet(B)_2 (absent)",
                               "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)",
                               "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)",
                               "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)",
                               "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)",
                               "floR_2  (present)", "floR_2 (absent)",
                               "sul2_2 (present)", "sul2_2 (absent)",
                               "dfrA1_8 (present)", "dfrA1_8 (absent)",
                               "mph(A)_2 (present)", "mph(A)_2 (absent)", 
                               "qnrA1_1 (present)", "qnrA1_1 (absent)"),
                    name="Serovar -> BAPS1 -> ST -> cgMLST -> AMR loci") +  # add colors and labels for the scale 
  theme(legend.title=element_text(size=24, face = "bold"), 
    legend.text=element_text(size=22)) +
        theme(legend.position = "none")  # customize figure's title, legend, font
figure_5
```
Figure 5 with legend. Phylogeny-guided mapping of hierarchical genotypes along with ST-differentiating AMR loci.

```{r, include = TRUE}
# enter the ST data 
mlst <- read_csv('~/Documents/jove_paper/data/salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest and rename -/? to NAs
data1 <- mlst %>%
          select(id2, ST) %>% # select columns 
            mutate_all(na_if, "-") %>% # change - to NA
              mutate_all(na_if, "?") %>% # change ? to NA
                  rename(id = id2) # rename column 
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data1$st <- ifelse(data1$ST == 5, "ST5", # aggregate minor STs as Others 
                   ifelse(data1$ST == 31, "ST31", 
                          ifelse(data1$ST == 45, "ST45", 
                                 ifelse(data1$ST == 46, "ST46", 
                                        ifelse(data1$ST == 118, "ST118", 
                                               ifelse(data1$ST == 132, "ST132", 
                          ifelse(data1$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data1 <- data1 %>% select(-ST)
##########################################--------############################################
#########################################--------#############################################
# enter the Resfinder loci databases (AMR loci)
abx <- read_csv('~/Documents/jove_paper/data/sabricate_resfinder_output.csv')
# create the id column 
abx$id <- sapply(strsplit(as.character(abx$'#FILE'),'_'), "[", 1)
# change the name of the GENE column 
abx <- abx %>% mutate(gene = GENE)
# select columns of interest
abx1 <- abx %>% select(id, gene)
##########################################--------############################################
#########################################--------#############################################
# join ST and AMR loci data 
d1 <- left_join(data1, abx1, on = "id")
# check for NAs or missing values
skim(d1)
# group ST missing values as "Other STs"
d1 <- d1 %>% mutate(st = replace_na(st, "Other STs"))
# check for NAs again
skim(d1)
##########################################--------############################################
#########################################--------#############################################
# calculate the proportion of AMR loci for each ST lineage
# calculations for ST5
d2a <- d1 %>% filter(st == "ST5")
# for ST5, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3a <- d2a %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4a <- d3a %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[8]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5a <- d4a %>% mutate(st = "ST5")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST31
d2b <- d1 %>% filter(st == "ST31")
# for ST31, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3b <- d2b %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4b <- d3b %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[4]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5b <- d4b %>% mutate(st = "ST31")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST45
d2c <- d1 %>% filter(st == "ST45")
# for ST45, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3c <- d2c %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4c <- d3c %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[6]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5c <- d4c %>% mutate(st = "ST45")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST46
d2d <- d1 %>% filter(st == "ST46")
# for ST46, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3d <- d2d %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4d <- d3d %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[7]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5d <- d4d %>% mutate(st = "ST46")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST118
d2e <- d1 %>% filter(st == "ST118")
# for ST118, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3e <- d2e %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations 
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4e <- d3e %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[2]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5e <- d4e %>% mutate(st = "ST118")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST132
d2f <- d1 %>% filter(st == "ST132")
# for ST132, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3f <- d2f %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4f <- d3f %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[3]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5f <- d4f %>% mutate(st = "ST132")
##########################################--------############################################
#########################################--------#############################################
# calculations for ST350
d2g <- d1 %>% filter(st == "ST350")
# for ST350, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3g <- d2g %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4g <- d3g %>% 
         group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[5]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5g <- d4g %>% mutate(st = "ST350")
##########################################--------############################################
#########################################--------#############################################
# calculations for Other ST lineages
d2h <- d1 %>% filter(st == "Other STs")
# for other STs, calculate the proportion of AMR loci and only keep proportion greater than 1%
d3h <- d2h %>% 
    select(id, gene) %>% # select columns 
    group_by(id, gene) %>% # group by columns 
    summarize(count = n()) %>% # count observations
    mutate(count = replace(count, count == 2, 1)) %>% # replace values in a column 
    filter(count <= 1) # filter based on a threshold 
d4h <- d3h %>% group_by(gene) %>% # group by column 
         summarize(value = n()) %>% # count observations 
         mutate(total = table(data1$st)[1]) %>% # get the total number of observations per ST
         mutate(prop = (value/total)*100) # calculate proportions 
d5h <- d4h %>% mutate(st = "Other STs")
##########################################--------############################################
#########################################--------#############################################
# combined datasets
d6 <- rbind(d5a, d5b, d5c, d5d, d5e, d5f, d5g, d5h)
##########################################--------############################################
#########################################--------#############################################
# filter AMR loci with proportion higher than or equal to %
d7 <- d6 %>% filter(prop >= 10)
# create a list of genes with proportion >= 10%
d8 <- d7 %>% select(gene) %>% unique()
# transform dataframe to list of characters or vector
d9 <- pull(d8, gene)
##########################################--------############################################
#########################################--------#############################################
# spread the abx1 table - from long to wide format
# count gene occurrences
abx2 <- abx1 %>% 
    group_by(id, gene) %>% # group by id and gene
    summarize(count = n()) %>% # count the number of observations 
    filter(count <= 1) # filter genes with 0 or 1 counts 
# spread from long to wide format
abx3 <- spread(abx2, key = gene, value = count)
# select columns of interest
abx4 <- abx3 %>% select(d9)
# replace all NAs with zeros
abx4[is.na(abx4)] <- 0
# merge with original data containing the hierarchical genotypes
d4 <- left_join(d3, abx4, on = "id")
##########################################--------############################################
#########################################--------#############################################
# change binary values from AMR data locus by locus
d4 <- d4 %>% mutate(`aac(6')-Iaa_1` = factor(ifelse(`aac(6')-Iaa_1` == 1, "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)")))
d4 <- d4 %>% mutate(`mdf(A)_1` = factor(ifelse(`mdf(A)_1` == 1, "mdf(A)_1 (present)", "mdf(A)_1 (absent)")))
d4 <- d4 %>% mutate(aadA2_1 = factor(ifelse(aadA2_1 == 1, "aadA2_1 (present)", "aadA2_1 (absent)")))
d4 <- d4 %>% mutate(`ant(3'')-Ia_1` = factor(ifelse(`ant(3'')-Ia_1` == 1, "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_1` = factor(ifelse(`aph(3')-Ia_1` == 1, "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)")))
d4 <- d4 %>% mutate(`blaCARB-2_1` = factor(ifelse(`blaCARB-2_1` == 1, "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)")))
d4 <- d4 %>% mutate(catA1_1 = factor(ifelse(catA1_1 == 1, "catA1_1 (present)", "catA1_1 (absent)")))
d4 <- d4 %>% mutate(sul1_5 = factor(ifelse(sul1_5 == 1, "sul1_5 (present)", "sul1_5 (absent)")))
d4 <- d4 %>% mutate(`tet(A)_6` = factor(ifelse(`tet(A)_6` == 1, "tet(A)_6 (present)", "tet(A)_6 (absent)")))
d4 <- d4 %>% mutate(`tet(B)_2` = factor(ifelse(`tet(B)_2` == 1, "tet(B)_2 (present)", "tet(B)_2 (absent)")))
d4 <- d4 %>% mutate(`aph(3'')-Ib_5` = factor(ifelse(`aph(3'')-Ib_5` == 1, "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)")))
d4 <- d4 %>% mutate(`aph(3')-Ia_9` = factor(ifelse(`aph(3')-Ia_9` == 1, "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)")))
d4 <- d4 %>% mutate(`aph(6)-Id_1` = factor(ifelse(`aph(6)-Id_1` == 1, "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)")))
d4 <- d4 %>% mutate(`blaCMY-2_1` = factor(ifelse(`blaCMY-2_1` == 1, "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)")))
d4 <- d4 %>% mutate(floR_2 = factor(ifelse(floR_2 == 1, "floR_2  (present)", "floR_2 (absent)")))
d4 <- d4 %>% mutate(sul2_2 = factor(ifelse(sul2_2 == 1, "sul2_2 (present)", "sul2_2 (absent)")))
d4 <- d4 %>% mutate(dfrA1_8 = factor(ifelse(dfrA1_8 == 1, "dfrA1_8 (present)", "dfrA1_8 (absent)")))
d4 <- d4 %>% mutate(`mph(A)_2` = factor(ifelse(`mph(A)_2` == 1, "mph(A)_2 (present)", "mph(A)_2 (absent)")))
d4 <- d4 %>% mutate(qnrA1_1 = factor(ifelse(qnrA1_1 == 1, "qnrA1_1 (present)", "qnrA1_1 (absent)")))
##########################################--------############################################
#########################################--------#############################################
# enter the phylogeny data
tree <- read.tree("~/Documents/jove_paper/data/newport_phylogeny.tree")
# bring data containing hierarchical genotypes and AMR loci
d5 <- d4
# to plot with the phylogeny using ggtree we need to make the id column into index first 
d6 <- column_to_rownames(d5, var = "id") 
# create the tree
# adjusting the parameters to make the tree visible or however you wish requires some trial and error 
tree_plot <- ggtree(tree, layout = "circular") + xlim(-50, NA)
# plot figure 1 - color scheme for each layer of the plot should be chosen based on the user preferences 
figure_5b <- gheatmap(tree_plot, d6, offset=.0, width=50, colnames = FALSE) +  # visualize the tree with metadata
  scale_fill_manual(values = c("coral", "darkblue",
                               "cornflowerblue", "coral", "purple", "red", "brown", "darkseagreen3", "darkblue", "darkgreen", "yellow",
                               "orange", "darkblue", "purple", "darkgreen", "darkred", "steelblue", "black", "coral",
                               "black", "darkgreen", "purple", "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray",
                               "darkblue", "gray"),
                    breaks = c("Newport", "Other serovars",
                               "BAPS1 sub-group 1", "BAPS1 sub-group 2", "BAPS1 sub-group 3", "BAPS1 sub-group 4", 
                               "BAPS1 sub-group 5", "BAPS1 sub-group 6", "BAPS1 sub-group 7", "BAPS1 sub-group 8", 
                               "BAPS1 sub-group 9",
                               "ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs",
                               "cgMLST 1468400426", "cgMLST 1271156802", "cgMLST 3336043520", "cgMLST 88443731", "Other cgMLSTs",
                               "aac(6')-Iaa_1 (present)", "aac(6')-Iaa_1 (absent)",
                               "mdf(A)_1 (present)", "mdf(A)_1 (absent)",
                               "aadA2_1 (present)", "aadA2_1 (absent)",
                               "ant(3'')-Ia_1 (present)", "ant(3'')-Ia_1 (absent)",
                               "aph(3')-Ia_1 (present)", "aph(3')-Ia_1 (absent)",
                               "blaCARB-2_1 (present)", "blaCARB-2_1 (absent)",
                               "catA1_1 (present)", "catA1_1 (absent)",
                               "sul1_5 (present)", "sul1_5 (absent)", 
                               "tet(A)_6 (present)", "tet(A)_6 (absent)", 
                               "tet(B)_2 (present)", "tet(B)_2 (absent)",
                               "aph(3'')-Ib_5 (present)", "aph(3'')-Ib_5 (absent)",
                               "aph(3')-Ia_9 (present)", "aph(3')-Ia_9 (absent)",
                               "aph(6)-Id_1 (present)", "aph(6)-Id_1 (absent)",
                               "blaCMY-2_1 (present)", "blaCMY-2_1 (absent)",
                               "floR_2  (present)", "floR_2 (absent)",
                               "sul2_2 (present)", "sul2_2 (absent)",
                               "dfrA1_8 (present)", "dfrA1_8 (absent)",
                               "mph(A)_2 (present)", "mph(A)_2 (absent)", 
                               "qnrA1_1 (present)", "qnrA1_1 (absent)"),
                    name="Serovar -> BAPS1 -> ST -> cgMLST -> AMR loci") +  # add colors and labels for the scale 
  theme(legend.title=element_text(size=24, face = "bold"), 
    legend.text=element_text(size=22)) # customize figure's title, legend, font
figure_5b
```
