---
title: "Supplementary Figure S3"
author: "Joao Carlos Gomes-Neto"
date: "8/2/2021"
output: html_document
---

When beginning running this script, you can remove the # that comes before the install.packages() function, but when running it the second time around, and subsequently, comment the function out using #. That way you avoid creating issues with dependencies, and different versions of the program. 

```{r, include = TRUE}
# install tidyverse 
# install.packages("tidyverse")
# install skimr 
# install.packages("skimr")
# install vegan 
# install.packages("vegan")
# install forcats
# install.packages("forcats")
# install naniar
# install.packages("naniar")
# install ggpubr
# install.packages("ggpubr")
# install ggrepel
# install.packages("ggrepel")
# install reshape2
# install.packages("reshape2")
# install RColorBrewer
# install.packages("RColorBrewer")
# install ggtree 
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ggtree")
# installation of ggtree will prompt a question about installation - answer is "a" to install/update all dependencies 
```

How to activate packages prior to utilization.

```{r, include = TRUE}
# Load previously installed packages
library(tidyverse)
library(skimr)
library(vegan)
library(forcats)
library(naniar)
library(ggtree)
library(ggpubr)
library(ggrepel)
library(reshape2)
library(RColorBrewer)
```

Enter and quality control all genotypic data including: serovar-predictions (generated by SISTR), BAPS level 1 (generated by fastbaps), ST lineages (generated by mlst), 
and cgMLST variants (generated by SISTR). All input files were generated by the describe programs which are part of the computational platform called ProkEvo.

```{r, include = TRUE}
# enter the BAPS data 
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# check the first six rows of the dataset
head(baps)
# changing the first two column names because for hierarchical analysis we only use BAPS1
colnames(baps)[1:2] <- c("id", "baps1") 
# check the first six rows of the datset again to see the change in column names
head(baps)
# select columns id and baps_1
baps1 <- baps %>% 
              select(id, baps1)
# check the first six rows of the dataset
head(baps1)
# quality control baps1 data 
skim(baps1)
# using a plotting strategy to check for missing values
vis_miss(baps1)
# rename BAPS level 1 sub-groups or haplotypes 
baps1$baps_1 <- ifelse(baps1$baps1 == 1, "BAPS1 sub-group 1", # the ifelse functions tests for different conditions prior to assigning groups to one category or another 
                   ifelse(baps1$baps1 == 2, "BAPS1 sub-group 2", 
                          ifelse(baps1$baps1 == 3, "BAPS1 sub-group 3", 
                                 ifelse(baps1$baps1 == 4, "BAPS1 sub-group 4", 
                                        ifelse(baps1$baps1 == 5, "BAPS1 sub-group 5", 
                                               ifelse(baps1$baps1 == 6, "BAPS1 sub-group 6", 
                                                       ifelse(baps1$baps1 == 7, "BAPS1 sub-group 7", 
                                                               ifelse(baps1$baps1 == 8, "BAPS1 sub-group 8", "BAPS1 sub-group 9"))))))))
##########################################--------############################################
#########################################--------#############################################
# enter MLST results
mlst <- read_csv('salmonellast_output.csv')
# check the first six rows of the dataset
head(mlst)
# generate the id column by deriving it from the FILE column
mlst$id <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# check the first six rows of the dataset
head(mlst)
# select the id and ST columns 
mlst1 <- mlst %>%
          select(id, ST) # select id and ST columns 
# use the table() function to detect extraneous characters such as "-" or "?" in the data - those are ST misclassifications
# check for the presence of ST misclassification in the ST column 
table(mlst1$ST)
# check for missing values
skim(mlst1)
# mutate the "-" character to NA (missing value)
mlst1 <- mlst1 %>%
            mutate_all(na_if, "-") # fill all - with NAs
# check the first six rows of the dataset
head(mlst1)
# quality control baps1 data 
skim(mlst1)
# using a plotting strategy to check for missing values
vis_miss(mlst1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# Enter SISTR results
sistr <- read_csv('sistr_output.csv')
# check the first six rows of the dataset
head(sistr)
# generate the id column by deriving it from the genome column
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select the id and cgmlst_ST columns 
sistr1 <- sistr %>%
            select(id, serovar_cgmlst, cgmlst_ST) # select id, serovar_cgmlst, and cgmlst_ST columns 
# check the first six rows of the dataset
head(sistr1)
# quality control baps1 data 
skim(sistr1)
# using a plotting strategy to check for missing values
vis_miss(sistr1)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# combine baps1, mlst1, and sistr1 datasets
d1 <- left_join(baps1, mlst1, on = "id") # merge datasets on identical ids 
d2 <- left_join(d1, sistr1, on = "id") # merge datasets on identical ids 
# check data dimensionality 
dim(d2)
# check the first six rows of the dataset
head(d2)
# quality control baps1 data 
skim(d2)
# using a plotting strategy to check for missing values
vis_miss(d2)
# At this stage missing values are not removed or dealt with until datasets are merged 
##########################################--------############################################
#########################################--------#############################################
# group all genomes not classified as Newport as "Other serovars"
# create a new column called serovar that contains the binary classification to group all SISTR-based misclassified genomes
d2$serovar <- ifelse(d2$serovar_cgmlst == "Newport", "Newport",
                            "Other serovars") # classify serovar_cgmlst into Newport or Others 
# check the first six rows of the dataset
head(d2)
# check the composition of the serovar column 
table(d2$serovar)
##########################################--------############################################
#########################################--------#############################################
# no transformation is needed for the baps1 dataset - no groupings or aggregations are needed 
##########################################--------############################################
#########################################--------#############################################
# check ST distribution to focus on major STs for all subsequent analyses by grouping by ST and counting
st_dist <- d2 %>%
  group_by(ST) %>% # group by the ST column
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange the counts in decreasing order 
st_dist
# based on the frequency analysis, the following STs were not aggregrated: ST118, ST45, ST5, ST132, ST31, ST350, and ST46
# create a new st column
d2$st <- ifelse(d2$ST == 5, "ST5", # create a new ST column for which minor STs are aggregated as Others 
                   ifelse(d2$ST == 31, "ST31", 
                          ifelse(d2$ST == 45, "ST45", 
                                 ifelse(d2$ST == 46, "ST46", 
                                        ifelse(d2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(d2$ST == 350, "ST350", "Other STs")))))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# check cgMLST distribution by ST to focus on major cgMLSTs for all subsequent analyses by grouping by STs and cgMLSTs and counting
cgmlst_dist <- d2 %>%
  group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
  count() %>% # count the number of observations 
  arrange(desc(n)) # arrange in descending order 
cgmlst_dist
# for the purposes of this analysis only the top four most frequent cgMLSTs were selected and respectively renamed
d2 <- mutate(d2, cgmlst = ifelse(cgmlst_ST %in% 1468400426, "cgMLST 1468400426", # create a new cgMLST column while aggregating minor cgMLST variants 
                                     ifelse(cgmlst_ST %in% 88443731, "cgMLST 88443731", 
                                            ifelse(cgmlst_ST %in% 1271156802, "cgMLST 1271156802",
                                                 ifelse(cgmlst_ST %in% 3336043520, "cgMLST 3336043520",  
                                            "Other cgMLSTs")))))
# check the first six rows of the dataset
head(d2)
##########################################--------############################################
#########################################--------#############################################
# select only needed columns for all subsequent analyses 
d3 <- d2 %>% 
          select(id, serovar, baps_1, st, cgmlst) # select columns of interest which are described within parenthesis 
# check data dimensionality to make sure it matches that of d2 
dim(d3)
# check the first six rows of the dataset
head(d3)
# quality control baps1 data 
skim(d3)
# using a plotting strategy to check for missing values
vis_miss(d3)
# make sure there is no NA in the dataset
sum(is.na(d3))
# replace NA in the ST column
d3 <- d3 %>% mutate(st = replace_na(st, "Other STs")) 
# make sure there is no NA in the dataset
sum(is.na(d3))
# check the distribution of serovars, baps1, st, and cgmlst classifications
table(d3$serovar)
table(d3$baps_1)
table(d3$st)
table(d3$cgmlst)
# check the first six rows in the dataset 
head(d3)
# check the last six rows in the dataset 
tail(d3)
# check for missing values
vis_miss(d3)
# check the data characteristics 
str(d3)
```

Supplementary Figure S3. Simpson's D diversity analysis of ST lineages using cgMLST and BAPS levels 1-6. 
Note - only genomes classified as S. Newport by SISTR within ProkEvo are included in this analysis.

```{r, include = TRUE}
# enter BAPS1-6 data
baps <- read_csv('fastbaps_partition_baps_prior_l6.csv')
# changing column names 
colnames(baps)[1:7] <- c("id", "BAPS1", "BAPS2", "BAPS3", "BAPS4", "BAPS5", "BAPS6") 
# assign baps data to d1
data1 <- baps
# transform BAPS sub-groups to factor (categorical data)
data1 <- data1 %>% mutate(BAPS1 = as.factor(BAPS1))
data1 <- data1 %>% mutate(BAPS2 = as.factor(BAPS2))
data1 <- data1 %>% mutate(BAPS3 = as.factor(BAPS3))
data1 <- data1 %>% mutate(BAPS4 = as.factor(BAPS4))
data1 <- data1 %>% mutate(BAPS5 = as.factor(BAPS5))
data1 <- data1 %>% mutate(BAPS6 = as.factor(BAPS6))
##########################################--------############################################
#########################################--------#############################################
# enter the ST data 
mlst <- read_csv('salmonellast_output.csv')
# generate the id column 
mlst$id2 <- sapply(strsplit(as.character(mlst$FILE),'_'), "[", 1)
# select columns of interest
data2 <- mlst %>%
          select(id2, ST) %>% # select columns of interest
            mutate_all(na_if, "-") %>% # transform any - to NA
              mutate_all(na_if, "?") %>% # transform any ? to NA
                  rename(id = id2) # rename id2 column to id
# categorize ST data to aggregate minor STs and keep only major ST lineages separately 
data2$st <- ifelse(data2$ST == 5, "ST5", # group major STs separately, and minor STs as Others 
                   ifelse(data2$ST == 31, "ST31", 
                          ifelse(data2$ST == 45, "ST45", 
                                 ifelse(data2$ST == 46, "ST46", 
                                        ifelse(data2$ST == 118, "ST118", 
                                               ifelse(d2$ST == 132, "ST132", 
                          ifelse(data2$ST == 350, "ST350", "Other STs")))))))
# filter out the numerical ST column 
data2 <- data2 %>% select(-ST) 
##########################################--------############################################
#########################################--------#############################################
# enter SISTR results
sistr <- read_csv('sistr_output.csv')
# generate the id column 
sistr$id <- sapply(strsplit(as.character(sistr$genome),'_'), "[", 1)
# select id and cgmlst_ST columns 
data3 <- sistr %>%
            select(id, serovar, cgmlst_ST) %>% # select columns of interest
            mutate_all(na_if, "-") %>% # transform any - to NA
              mutate_all(na_if, "?") # transform any ? to NA
# group serovars as Newport or others
data3$serovar <- ifelse(data3$serovar == "Newport", "Newport",
                            "Other serovars") # group genomes to either Newport or other serovars 
##########################################--------############################################
#########################################--------#############################################
# merge datasets
data4 <- left_join(data1, data2, on = "id") # join datasets based on id
data5 <- left_join(data3, data4, on = "id") # join datasets based on id
# filter data for Newport only
data6 <- data5 %>% filter(serovar == "Newport")
# check for missing values
skim(data6)
# group ST missing values as "Other STs"
data6 <- data6 %>% mutate(st = replace_na(st, "Other STs"))
# check the distribution of serovars to make sure there is only Newport
table(data6$serovar)
##########################################--------############################################
#########################################--------#############################################
# cgMLST diversity 
# drop the STs and cgMLSTs with NAs, group by ST and cgMLST and then calculate Simpson's index
cgmlst_div <- data6 %>%
              drop_na(st, cgmlst_ST) %>% # drop NAs 
              mutate(cgmlst_ST = as.factor(cgmlst_ST)) %>% # transform column to categorical 
              select(st, cgmlst_ST) %>% # select columns of interest 
              group_by(st, cgmlst_ST) %>% # group by st and cgmlst_ST
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate the index of diversity
              group_by(st) %>% # group by st 
              summarise(simpson = mean(simpson)) %>% # get the mean of the simpson index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # convert into long format
              mutate(strat = "cgMLST") # create a strat column
##########################################--------############################################
#########################################--------#############################################
# BAPS based diversity (calculated across BAPS levels 1 through 6) 
# BAPS level 1
# drop the STs and BAPS1 with NAs, group by ST and BAPS1 and then calculate Simpson's index
baps1 <- data6 %>%
              select(st, BAPS1) %>% # select columns 
              drop_na(st, BAPS1) %>% # drop NAs 
              group_by(st, BAPS1) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format 
              mutate(strat = "BAPS1") # create a strat column 
# BAPS level 2
# drop the STs and BAPS2 with NAs, group by ST and BAPS2 and then calculate Simpson's index
baps2 <- data6 %>%
              select(st, BAPS2) %>%  # select columns 
              drop_na(st, BAPS2) %>% # drop NAs 
              group_by(st, BAPS2) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS2") # create a strat column 
# BAPS level 3
# drop the STs and BAPS3 with NAs, group by ST and BAPS3 and then calculate Simpson's index
baps3 <- data6 %>%
              select(st, BAPS3) %>% # select columns 
              drop_na(st, BAPS3) %>% # drop NAs 
              group_by(st, BAPS3) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS3") # create a strat column 
# BAPS level 4
# drop the STs and BAPS4 with NAs, group by ST and BAPS4 and then calculate Simpson's index
baps4 <- data6 %>%
              select(st, BAPS4) %>% # select columns 
              drop_na(st, BAPS4) %>% # drop NAs 
              group_by(st, BAPS4) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS4") # create a strat column 
# BAPS level 5
# drop the STs and BAPS5 with NAs, group by ST and BAPS5 and then calculate Simpson's index
baps5 <- data6 %>%
              select(st, BAPS5) %>% # select columns 
              drop_na(st, BAPS5) %>% # drop NAs 
              group_by(st, BAPS5) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS5") # create a strat column 
# BAPS level 6
# drop the STs and BAPS6 with NAs, group by ST and BAPS6 and then calculate Simpson's index
baps6 <- data6 %>%
              select(st, BAPS6) %>%  # select columns 
              drop_na(st, BAPS6) %>% # drop NAs 
              group_by(st, BAPS6) %>% # group by columns 
              summarise(n = n()) %>% # count observations 
              mutate(simpson = diversity(n, "simpson")) %>% # calculate diversity 
              group_by(st) %>% # group by column 
              summarise(simpson = mean(simpson)) %>% # calculate the mean of the index 
              melt(id.vars=c("st"), measure.vars="simpson",
                    variable.name="index", value.name="value") %>% # covert into long format
              mutate(strat = "BAPS6") # create a strat column 
##########################################--------############################################
#########################################--------#############################################
# Concatenate all datasets
data7 <- rbind(cgmlst_div, baps1, baps2, baps3, baps4, baps5, baps6) 
# order the ST column
data7$st <- factor(data7$st, levels=c("ST5", "ST31", "ST45", "ST46", "ST118", "ST132", "ST350", "Other STs"))
##########################################--------############################################
#########################################--------#############################################
# plot Simpson's diversity analysis
# order the strat colunmn 
data7$strat <- factor(data7$strat, levels=c("cgMLST", "BAPS6", "BAPS5", "BAPS4", "BAPS3", "BAPS2", "BAPS1"))
sup_fig_3 <- ggplot(data7, aes(x = strat, y = value)) +  # show strata on x-axis and index values on y-axis
  xlab("") + ylab("Index value") + ylim(0,1) +  # set labels for axis and limit for y-axis
  theme_bw() + # set plot background 
  theme(axis.text.y = element_text(size = 28)) + # change y-axis text font size
  theme(axis.title.y = element_text(size = 30, face = "bold")) + # change y-axis title font size and face
  theme(axis.title.x = element_text(size = 30, face = "bold")) + # change x-axis title font size and face
  theme(axis.text.x = element_text(angle = 70, hjust = 1, size = 24)) + # change x-axis text font size, angle, and orientation 
  theme(strip.text.x = element_text(size = 30, colour = "black", angle = 0)) +  # customize figure's title, legend, font
  geom_col(fill = "steelblue") +  # fill the bars that represent the values with blue
  facet_wrap(~st) +  # generate multi-plots for STs
  coord_flip()  # flip x and y axis
sup_fig_3
```

